// igniteui-cli/templates/react/es6/grid/grid-custom/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
//TODO:
const GridHelper_1 = require("../../../../../lib/project-utility/GridHelper");
const ReactTemplate_1 = require("../../../../../lib/templates/ReactTemplate");
class GridCustomTemplate extends ReactTemplate_1.ReactTemplate {
    /**
     *
     */
    constructor() {
        super(__dirname);
        this.extraConfigurations = [];
        this.userExtraConfiguration = {};
        this.id = "grid-custom";
        this.name = "Custom Grid";
        this.widget = "igGrid";
        this.description = "igGrid custom template for React";
        this.projectType = "es6";
        this.components = ["Grid"];
        this.controlGroup = "Data Grids";
        this.dependencies = ["igGrid"];
        this.gridHelper = new GridHelper_1.GridHelper();
        this.hasExtraConfiguration = true;
        this.extraConfigurations.push({
            choices: [
                "Sorting", "Selection", "Updating", "Filtering", "ColumnMoving",
                "GroupBy", "Summaries", "Resizing", "Hiding"
            ],
            default: "",
            key: "features",
            message: "Select features for the igGrid",
            type: 1 /* MultiChoice */
        });
    }
    generateFiles(projectPath, name, ...options) {
        const features = this.gridHelper.generateFeatures(this.userExtraConfiguration["features"], 5);
        const config = { "$(gridfeatures)": features };
        return super.generateFiles(projectPath, name, { extraConfig: config });
    }
    getExtraConfiguration() {
        return this.extraConfigurations;
    }
    setExtraConfiguration(extraConfigKeys) {
        this.userExtraConfiguration = extraConfigKeys;
    }
}
module.exports = new GridCustomTemplate();
