// igniteui-cli/templates/jquery/js/tree-grid/tree-grid-export/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const GridHelper_1 = require("../../../../../lib/project-utility/GridHelper");
const jQueryTemplate_1 = require("../../../../../lib/templates/jQueryTemplate");
class TreeGridExportTemplate extends jQueryTemplate_1.jQueryTemplate {
    /**
     *
     */
    constructor() {
        super(__dirname);
        this.userExtraConfiguration = {};
        this.extraConfigurations = [];
        this.id = "tree-grid-export";
        this.name = "TreeGrid Export";
        this.description = "igTreeGrid exporting template";
        this.projectType = "js";
        this.components = ["Tree Grid"];
        this.controlGroup = "Data Grids";
        this.dependencies = ["igTreeGrid", "igExcel", "igGridExcelExporter"];
        this.gridHelper = new GridHelper_1.GridHelper();
        this.gridHelper.tree = true;
        this.hasExtraConfiguration = true;
        const featureConfiguration = {
            choices: ["Filtering", "Hiding"],
            default: "",
            key: "features",
            message: "Select optional features for the export template",
            type: 1 /* MultiChoice */
        };
        this.extraConfigurations.push(featureConfiguration);
    }
    setExtraConfiguration(extraConfigKeys) {
        this.userExtraConfiguration = extraConfigKeys;
    }
    generateFiles(projectPath, name, ...options) {
        const features = this.gridHelper.generateFeatures(this.userExtraConfiguration["features"], 4);
        const config = { "$(treeGridFeatures)": features };
        return super.generateFiles(projectPath, name, { extraConfig: config });
    }
    getExtraConfiguration() {
        return this.extraConfigurations;
    }
}
module.exports = new TreeGridExportTemplate();
