// igniteui-cli/templates/jquery/js/grid/grid-export/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const GridHelper_1 = require("../../../../../lib/project-utility/GridHelper");
const jQueryTemplate_1 = require("../../../../../lib/templates/jQueryTemplate");
class GridExportTemplate extends jQueryTemplate_1.jQueryTemplate {
    /**
     *
     */
    constructor() {
        super(__dirname);
        this.userExtraConfiguration = {};
        this.extraConfigurations = [];
        this.id = "grid-export";
        this.name = "Grid Export";
        this.description = "igGrid export template";
        this.projectType = "js";
        this.components = ["Grid"];
        this.controlGroup = "Data Grids";
        this.dependencies = ["igGrid", "igExcel", "igGridExcelExporter"];
        this.gridHelper = new GridHelper_1.GridHelper();
        this.hasExtraConfiguration = true;
        const featureConfiguration = {
            choices: ["Summaries", "Hiding"],
            default: "",
            key: "features",
            message: "Select features for the igGrid",
            type: 1 /* MultiChoice */
        };
        this.extraConfigurations.push(featureConfiguration);
    }
    setExtraConfiguration(extraConfigKeys) {
        this.userExtraConfiguration = extraConfigKeys;
    }
    generateFiles(projectPath, name, ...options) {
        const features = this.gridHelper.generateFeatures(this.userExtraConfiguration["features"], 4);
        const config = { "$(gridfeatures)": features };
        return super.generateFiles(projectPath, name, { extraConfig: config });
    }
    getExtraConfiguration() {
        return this.extraConfigurations;
    }
}
module.exports = new GridExportTemplate();
