// igniteui-angular/esm5/lib/grids/hierarchical-grid/row-island-api.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
var IgxRowIslandAPIService = /** @class */ (function () {
    function IgxRowIslandAPIService() {
        this.change = new Subject();
        this.state = new Map();
        this.destroyMap = new Map();
        this.childRowIslands = new Map();
        this.childGrids = new Map();
    }
    /**
     * @param {?} rowIsland
     * @return {?}
     */
    IgxRowIslandAPIService.prototype.register = /**
     * @param {?} rowIsland
     * @return {?}
     */
    function (rowIsland) {
        this.state.set(rowIsland.id, rowIsland);
        this.destroyMap.set(rowIsland.id, new Subject());
    };
    /**
     * @param {?} rowIsland
     * @return {?}
     */
    IgxRowIslandAPIService.prototype.unsubscribe = /**
     * @param {?} rowIsland
     * @return {?}
     */
    function (rowIsland) {
        this.state.delete(rowIsland.id);
    };
    /**
     * @param {?} id
     * @return {?}
     */
    IgxRowIslandAPIService.prototype.get = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.state.get(id);
    };
    /**
     * @param {?} id
     * @return {?}
     */
    IgxRowIslandAPIService.prototype.unset = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        this.state.delete(id);
        this.destroyMap.delete(id);
    };
    /**
     * @param {?} oldId
     * @param {?} newId
     * @return {?}
     */
    IgxRowIslandAPIService.prototype.reset = /**
     * @param {?} oldId
     * @param {?} newId
     * @return {?}
     */
    function (oldId, newId) {
        /** @type {?} */
        var destroy = this.destroyMap.get(oldId);
        /** @type {?} */
        var rowIsland = this.get(oldId);
        this.unset(oldId);
        if (rowIsland) {
            this.state.set(newId, rowIsland);
        }
        if (destroy) {
            this.destroyMap.set(newId, destroy);
        }
    };
    /**
     * @param {?} rowIsland
     * @return {?}
     */
    IgxRowIslandAPIService.prototype.registerChildRowIsland = /**
     * @param {?} rowIsland
     * @return {?}
     */
    function (rowIsland) {
        this.childRowIslands.set(rowIsland.key, rowIsland);
        this.destroyMap.set(rowIsland.key, new Subject());
    };
    /**
     * @param {?} rowIsland
     * @return {?}
     */
    IgxRowIslandAPIService.prototype.unsetChildRowIsland = /**
     * @param {?} rowIsland
     * @return {?}
     */
    function (rowIsland) {
        this.childRowIslands.delete(rowIsland.key);
        this.destroyMap.delete(rowIsland.key);
    };
    /**
     * @param {?} rowIslandKey
     * @return {?}
     */
    IgxRowIslandAPIService.prototype.getChildRowIsland = /**
     * @param {?} rowIslandKey
     * @return {?}
     */
    function (rowIslandKey) {
        return this.childRowIslands.get(rowIslandKey);
    };
    /**
     * @param {?} parentRowID
     * @param {?} grid
     * @return {?}
     */
    IgxRowIslandAPIService.prototype.registerChildGrid = /**
     * @param {?} parentRowID
     * @param {?} grid
     * @return {?}
     */
    function (parentRowID, grid) {
        this.childGrids.set(parentRowID, grid);
    };
    /**
     * @param {?=} inDepth
     * @return {?}
     */
    IgxRowIslandAPIService.prototype.getChildGrids = /**
     * @param {?=} inDepth
     * @return {?}
     */
    function (inDepth) {
        /** @type {?} */
        var allChildren = [];
        this.childGrids.forEach(function (grid) {
            allChildren.push(grid);
        });
        if (inDepth) {
            this.childRowIslands.forEach(function (layout) {
                allChildren = allChildren.concat(layout.rowIslandAPI.getChildGrids(inDepth));
            });
        }
        return allChildren;
    };
    /**
     * @param {?} rowID
     * @return {?}
     */
    IgxRowIslandAPIService.prototype.getChildGridByID = /**
     * @param {?} rowID
     * @return {?}
     */
    function (rowID) {
        return this.childGrids.get(rowID);
    };
    return IgxRowIslandAPIService;
}());
export { IgxRowIslandAPIService };
if (false) {
    /** @type {?} */
    IgxRowIslandAPIService.prototype.rowIsland;
    /** @type {?} */
    IgxRowIslandAPIService.prototype.change;
    /**
     * @type {?}
     * @protected
     */
    IgxRowIslandAPIService.prototype.state;
    /**
     * @type {?}
     * @protected
     */
    IgxRowIslandAPIService.prototype.destroyMap;
    /**
     * @type {?}
     * @protected
     */
    IgxRowIslandAPIService.prototype.childRowIslands;
    /**
     * @type {?}
     * @protected
     */
    IgxRowIslandAPIService.prototype.childGrids;
}
