// igniteui-angular/esm5/lib/grids/drag-select.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Output, EventEmitter, ElementRef, NgZone } from '@angular/core';
import { interval, Subject } from 'rxjs';
import { filter, takeUntil } from 'rxjs/operators';
/** @enum {number} */
var DragScrollDirection = {
    NONE: 0,
    LEFT: 1,
    TOP: 2,
    RIGHT: 3,
    BOTTOM: 4,
    TOPLEFT: 5,
    TOPRIGHT: 6,
    BOTTOMLEFT: 7,
    BOTTOMRIGHT: 8,
};
export { DragScrollDirection };
DragScrollDirection[DragScrollDirection.NONE] = 'NONE';
DragScrollDirection[DragScrollDirection.LEFT] = 'LEFT';
DragScrollDirection[DragScrollDirection.TOP] = 'TOP';
DragScrollDirection[DragScrollDirection.RIGHT] = 'RIGHT';
DragScrollDirection[DragScrollDirection.BOTTOM] = 'BOTTOM';
DragScrollDirection[DragScrollDirection.TOPLEFT] = 'TOPLEFT';
DragScrollDirection[DragScrollDirection.TOPRIGHT] = 'TOPRIGHT';
DragScrollDirection[DragScrollDirection.BOTTOMLEFT] = 'BOTTOMLEFT';
DragScrollDirection[DragScrollDirection.BOTTOMRIGHT] = 'BOTTOMRIGHT';
var IgxGridDragSelectDirective = /** @class */ (function () {
    function IgxGridDragSelectDirective(ref, zone) {
        var _this = this;
        this.ref = ref;
        this.zone = zone;
        this.onDragStop = new EventEmitter();
        this.onDragScroll = new EventEmitter();
        this.end$ = new Subject();
        this.lastDirection = DragScrollDirection.NONE;
        this.startDragSelection = function (ev) {
            if (!_this.activeDrag) {
                return;
            }
            /** @type {?} */
            var x = ev.clientX;
            /** @type {?} */
            var y = ev.clientY;
            /** @type {?} */
            var direction = _this._measureDimensions(x, y);
            if (direction === _this.lastDirection) {
                return;
            }
            _this.unsubscribe();
            _this._sub = _this._interval$.subscribe(function () { return _this.onDragScroll.emit(direction); });
            _this.lastDirection = direction;
        };
        this.stopDragSelection = function () {
            if (!_this.activeDrag) {
                return;
            }
            _this.onDragStop.emit(false);
            _this.unsubscribe();
            _this.lastDirection = DragScrollDirection.NONE;
        };
        this._interval$ = interval(100).pipe(takeUntil(this.end$), filter(function () { return _this.activeDrag; }));
    }
    Object.defineProperty(IgxGridDragSelectDirective.prototype, "activeDrag", {
        get: /**
         * @return {?}
         */
        function () {
            return this._activeDrag;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            if (val !== this._activeDrag) {
                this.unsubscribe();
                this._activeDrag = val;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxGridDragSelectDirective.prototype, "nativeElement", {
        get: /**
         * @return {?}
         */
        function () {
            return this.ref.nativeElement;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxGridDragSelectDirective.prototype, "clientRect", {
        get: /**
         * @return {?}
         */
        function () {
            return this.nativeElement.getBoundingClientRect();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    IgxGridDragSelectDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.zone.runOutsideAngular(function () {
            _this.nativeElement.addEventListener('pointerover', _this.startDragSelection);
            _this.nativeElement.addEventListener('pointerleave', _this.stopDragSelection);
        });
    };
    /**
     * @return {?}
     */
    IgxGridDragSelectDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.zone.runOutsideAngular(function () {
            _this.nativeElement.removeEventListener('pointerover', _this.startDragSelection);
            _this.nativeElement.removeEventListener('pointerleave', _this.stopDragSelection);
        });
        this.unsubscribe();
        this.end$.complete();
    };
    /**
     * @param {?} x
     * @param {?} y
     * @return {?}
     */
    IgxGridDragSelectDirective.prototype._measureDimensions = /**
     * @param {?} x
     * @param {?} y
     * @return {?}
     */
    function (x, y) {
        /** @type {?} */
        var direction;
        /** @type {?} */
        var rect = this.clientRect;
        /** @type {?} */
        var RATIO = 0.15;
        /** @type {?} */
        var offsetX = Math.trunc(x - rect.left);
        /** @type {?} */
        var offsetY = Math.trunc(y - rect.top);
        /** @type {?} */
        var left = offsetX <= rect.width * RATIO;
        /** @type {?} */
        var right = offsetX >= rect.width * (1 - RATIO);
        /** @type {?} */
        var top = offsetY <= rect.height * RATIO;
        /** @type {?} */
        var bottom = offsetY >= rect.height * (1 - RATIO);
        if (top && left) {
            direction = DragScrollDirection.TOPLEFT;
        }
        else if (top && right) {
            direction = DragScrollDirection.TOPRIGHT;
        }
        else if (bottom && left) {
            direction = DragScrollDirection.BOTTOMLEFT;
        }
        else if (bottom && right) {
            direction = DragScrollDirection.BOTTOMRIGHT;
        }
        else if (top) {
            direction = DragScrollDirection.TOP;
        }
        else if (bottom) {
            direction = DragScrollDirection.BOTTOM;
        }
        else if (left) {
            direction = DragScrollDirection.LEFT;
        }
        else if (right) {
            direction = DragScrollDirection.RIGHT;
        }
        else {
            direction = DragScrollDirection.NONE;
        }
        return direction;
    };
    /**
     * @protected
     * @return {?}
     */
    IgxGridDragSelectDirective.prototype.unsubscribe = /**
     * @protected
     * @return {?}
     */
    function () {
        if (this._sub) {
            this._sub.unsubscribe();
        }
    };
    IgxGridDragSelectDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxGridDragSelect]'
                },] }
    ];
    /** @nocollapse */
    IgxGridDragSelectDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: NgZone }
    ]; };
    IgxGridDragSelectDirective.propDecorators = {
        activeDrag: [{ type: Input, args: ['igxGridDragSelect',] }],
        onDragStop: [{ type: Output }],
        onDragScroll: [{ type: Output }]
    };
    return IgxGridDragSelectDirective;
}());
export { IgxGridDragSelectDirective };
if (false) {
    /** @type {?} */
    IgxGridDragSelectDirective.prototype._activeDrag;
    /** @type {?} */
    IgxGridDragSelectDirective.prototype.onDragStop;
    /** @type {?} */
    IgxGridDragSelectDirective.prototype.onDragScroll;
    /**
     * @type {?}
     * @protected
     */
    IgxGridDragSelectDirective.prototype.end$;
    /**
     * @type {?}
     * @protected
     */
    IgxGridDragSelectDirective.prototype.lastDirection;
    /**
     * @type {?}
     * @protected
     */
    IgxGridDragSelectDirective.prototype._interval$;
    /**
     * @type {?}
     * @protected
     */
    IgxGridDragSelectDirective.prototype._sub;
    /** @type {?} */
    IgxGridDragSelectDirective.prototype.startDragSelection;
    /** @type {?} */
    IgxGridDragSelectDirective.prototype.stopDragSelection;
    /**
     * @type {?}
     * @private
     */
    IgxGridDragSelectDirective.prototype.ref;
    /**
     * @type {?}
     * @private
     */
    IgxGridDragSelectDirective.prototype.zone;
}
