// igniteui-angular/esm5/lib/grids/grid-column-resizer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ViewChild } from '@angular/core';
import { IgxColumnResizingService } from './grid-column-resizing.service';
import { IgxColumnResizerDirective } from './grid.common';
var IgxGridColumnResizerComponent = /** @class */ (function () {
    function IgxGridColumnResizerComponent(colResizingService) {
        this.colResizingService = colResizingService;
    }
    IgxGridColumnResizerComponent.decorators = [
        { type: Component, args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    preserveWhitespaces: false,
                    selector: 'igx-grid-column-resizer',
                    template: "<div class=\"igx-grid__th-resize-line\"\n    [style.left.px]=\"-99999\" igxResizer\n    [style.height.px]=\"colResizingService.resizerHeight\"\n    [restrictHResizeMax]=\"colResizingService.restrictResizeMax\"\n    [restrictHResizeMin]=\"colResizingService.restrictResizeMin\"\n    (resizeEnd)=\"colResizingService.resizeColumn($event)\">\n</div>"
                }] }
    ];
    /** @nocollapse */
    IgxGridColumnResizerComponent.ctorParameters = function () { return [
        { type: IgxColumnResizingService }
    ]; };
    IgxGridColumnResizerComponent.propDecorators = {
        resizer: [{ type: ViewChild, args: [IgxColumnResizerDirective, { static: true },] }]
    };
    return IgxGridColumnResizerComponent;
}());
export { IgxGridColumnResizerComponent };
if (false) {
    /** @type {?} */
    IgxGridColumnResizerComponent.prototype.resizer;
    /** @type {?} */
    IgxGridColumnResizerComponent.prototype.colResizingService;
}
