// igniteui-angular/esm5/lib/grids/igx-row-selectors.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, NgModule, TemplateRef } from '@angular/core';
/**
 * @hidden
 */
var IgxRowSelectorDirective = /** @class */ (function () {
    function IgxRowSelectorDirective(templateRef) {
        this.templateRef = templateRef;
    }
    IgxRowSelectorDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxRowSelector]'
                },] }
    ];
    /** @nocollapse */
    IgxRowSelectorDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return IgxRowSelectorDirective;
}());
export { IgxRowSelectorDirective };
if (false) {
    /** @type {?} */
    IgxRowSelectorDirective.prototype.templateRef;
}
/**
 * @hidden
 */
var IgxHeadSelectorDirective = /** @class */ (function () {
    function IgxHeadSelectorDirective(templateRef) {
        this.templateRef = templateRef;
    }
    IgxHeadSelectorDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxHeadSelector]'
                },] }
    ];
    /** @nocollapse */
    IgxHeadSelectorDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return IgxHeadSelectorDirective;
}());
export { IgxHeadSelectorDirective };
if (false) {
    /** @type {?} */
    IgxHeadSelectorDirective.prototype.templateRef;
}
/**
 * @hidden
 */
var IgxRowSelectorsModule = /** @class */ (function () {
    function IgxRowSelectorsModule() {
    }
    IgxRowSelectorsModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [IgxRowSelectorDirective, IgxHeadSelectorDirective],
                    exports: [IgxRowSelectorDirective, IgxHeadSelectorDirective]
                },] }
    ];
    return IgxRowSelectorsModule;
}());
export { IgxRowSelectorsModule };
