// igniteui-angular/esm2015/lib/services/transaction/transaction.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const TransactionType = {
    ADD: 'add',
    DELETE: 'delete',
    UPDATE: 'update',
};
export { TransactionType };
/**
 * @record
 */
export function Transaction() { }
if (false) {
    /** @type {?} */
    Transaction.prototype.id;
    /** @type {?} */
    Transaction.prototype.type;
    /** @type {?} */
    Transaction.prototype.newValue;
}
/**
 * \@experimental
 * @hidden
 * @record
 */
export function HierarchicalTransaction() { }
if (false) {
    /** @type {?} */
    HierarchicalTransaction.prototype.path;
}
/**
 * @record
 */
export function State() { }
if (false) {
    /** @type {?} */
    State.prototype.value;
    /** @type {?} */
    State.prototype.recordRef;
    /** @type {?} */
    State.prototype.type;
}
/**
 * \@experimental
 * @hidden
 * @record
 */
export function HierarchicalState() { }
if (false) {
    /** @type {?} */
    HierarchicalState.prototype.path;
}
/**
 * @record
 * @template T, S
 */
export function TransactionService() { }
if (false) {
    /**
     * Returns whether transaction is enabled for this service
     * @type {?}
     */
    TransactionService.prototype.enabled;
    /**
     * Event fired when transaction state has changed - add transaction, commit all transactions, undo and redo
     * @type {?|undefined}
     */
    TransactionService.prototype.onStateUpdate;
    /**
     * \@return if there are any transactions in the Undo stack
     * @type {?}
     */
    TransactionService.prototype.canUndo;
    /**
     * \@return if there are any transactions in the Redo stack
     * @type {?}
     */
    TransactionService.prototype.canRedo;
    /**
     * Adds provided  transaction with recordRef if any
     * @param {?} transaction Transaction to be added
     * @param {?=} recordRef Reference to the value of the record in the data source related to the changed item
     * @return {?}
     */
    TransactionService.prototype.add = function (transaction, recordRef) { };
    /**
     * Returns all recorded transactions in chronological order
     * @param {?=} id Optional record id to get transactions for
     * @return {?} All transaction in the service or for the specified record
     */
    TransactionService.prototype.getTransactionLog = function (id) { };
    /**
     * Remove the last transaction if any
     * @return {?}
     */
    TransactionService.prototype.undo = function () { };
    /**
     * Applies the last undone transaction if any
     * @return {?}
     */
    TransactionService.prototype.redo = function () { };
    /**
     * Returns aggregated changes from all transactions
     * @param {?} mergeChanges If set to true will merge each state's value over relate recordRef
     * and will record resulting value in the related transaction
     * @return {?} Collection of aggregated transactions for each changed record
     */
    TransactionService.prototype.getAggregatedChanges = function (mergeChanges) { };
    /**
     * Returns the state of the record with provided id
     * @param {?} id The id of the record
     * @param {?=} pending Should get pending state
     * @return {?} State of the record if any
     */
    TransactionService.prototype.getState = function (id, pending) { };
    /**
     * Returns value of the required id including all uncommitted changes
     * @param {?} id The id of the record to return value for
     * @param {?} mergeChanges If set to true will merge state's value over relate recordRef
     * and will return merged value
     * @return {?} Value with changes or **null**
     */
    TransactionService.prototype.getAggregatedValue = function (id, mergeChanges) { };
    /**
     * Applies all transactions over the provided data
     * @param {?} data Data source to update
     * @param {?=} id Optional record id to commit transactions for
     * @return {?}
     */
    TransactionService.prototype.commit = function (data, id) { };
    /**
     * Clears all transactions
     * @param {?=} id Optional record id to clear transactions for
     * @return {?}
     */
    TransactionService.prototype.clear = function (id) { };
    /**
     * Starts pending transactions. All transactions passed after call to startPending
     * will not be added to transaction log
     * @return {?}
     */
    TransactionService.prototype.startPending = function () { };
    /**
     * Clears all pending transactions and aggregated pending state. If commit is set to true
     * commits pending states as single transaction
     * @param {?} commit Should commit the pending states
     * @return {?}
     */
    TransactionService.prototype.endPending = function (commit) { };
}
