// igniteui-cli/templates/react/es6/hierarchical-grid/hierarchical-grid-custom/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ReactTemplate_1 = require("../../../../../lib/templates/ReactTemplate");
const GridHelper_1 = require("../../../../../lib/project-utility/GridHelper");
class HierarchicalGridCustomTemplate extends ReactTemplate_1.ReactTemplate {
    constructor() {
        super(__dirname);
        this.extraConfigurations = [];
        this.userExtraConfiguration = {};
        this.id = "hierarchical-grid-custom";
        this.widget = "igHierarchicalGrid";
        this.name = "Custom Hierarchical Grid";
        this.description = "igHierarchicalGrid custom template for React";
        this.projectType = "es6";
        this.components = ["Hierarchical Grid"];
        this.controlGroup = "Data Grids";
        this.dependencies = ["igHierarchicalGrid"];
        this.gridHelper = new GridHelper_1.GridHelper();
        this.gridHelper.hierarchical = true;
        this.hasExtraConfiguration = true;
        this.extraConfigurations.push({
            choices: [
                "Sorting", "Selection", "Updating", "Filtering", "ColumnMoving",
                "Resizing", "Hiding", "Paging", "Summaries"
            ],
            default: "",
            key: "features",
            message: "Select features for the igHierarchicalGrid",
            type: 1 /* MultiChoice */
        });
    }
    generateFiles(projectPath, name, ...options) {
        const features = this.gridHelper.generateFeatures(this.userExtraConfiguration["features"], 5);
        const config = { "$(gridfeatures)": features };
        return super.generateFiles(projectPath, name, { extraConfig: config });
    }
    getExtraConfiguration() {
        return this.extraConfigurations;
    }
    setExtraConfiguration(extraConfigKeys) {
        this.userExtraConfiguration = extraConfigKeys;
    }
}
module.exports = new HierarchicalGridCustomTemplate();
