// igniteui-angular/esm5/lib/drop-down/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { IgxDropDownComponent } from './drop-down.component';
import { IgxDropDownItemComponent } from './drop-down-item.component';
import { IgxDropDownItemNavigationDirective } from './drop-down-navigation.directive';
import { CommonModule } from '@angular/common';
import { IgxToggleModule } from '../directives/toggle/toggle.directive';
import { IgxSelectionAPIService } from '../core/selection';
import { IgxDropDownGroupComponent } from './drop-down-group.component';
export { IgxDropDownComponent } from './drop-down.component';
export { IgxDropDownItemComponent } from './drop-down-item.component';
export { IgxDropDownItemNavigationDirective } from './drop-down-navigation.directive';
export { IgxDropDownBase } from './drop-down.base';
export { IgxDropDownItemBase } from './drop-down-item.base';
export { IgxDropDownGroupComponent } from './drop-down-group.component';
/**
 * @hidden
 */
var IgxDropDownModule = /** @class */ (function () {
    function IgxDropDownModule() {
    }
    IgxDropDownModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [IgxDropDownComponent, IgxDropDownItemComponent, IgxDropDownGroupComponent,
                        IgxDropDownItemNavigationDirective],
                    exports: [IgxDropDownComponent, IgxDropDownItemComponent, IgxDropDownGroupComponent,
                        IgxDropDownItemNavigationDirective],
                    imports: [CommonModule, IgxToggleModule],
                    providers: [IgxSelectionAPIService]
                },] }
    ];
    return IgxDropDownModule;
}());
export { IgxDropDownModule };
