// igniteui-angular/esm5/lib/select/select.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { IgxInputDirective, IgxInputState } from './../directives/input/input.directive';
import { Component, ContentChildren, forwardRef, QueryList, ViewChild, Input, ContentChild, HostBinding, Directive, TemplateRef, ElementRef, ChangeDetectorRef, Optional, Injector, Inject } from '@angular/core';
import { NG_VALUE_ACCESSOR, NgControl } from '@angular/forms';
import { IgxInputGroupComponent } from '../input-group/input-group.component';
import { IgxDropDownComponent } from './../drop-down/drop-down.component';
import { IgxSelectItemComponent } from './select-item.component';
import { SelectPositioningStrategy } from './select-positioning-strategy';
import { AbsoluteScrollStrategy } from '../services/index';
import { IGX_DROPDOWN_BASE } from '../drop-down/drop-down.common';
import { IgxLabelDirective } from '../directives/label/label.directive';
import { IgxSelectionAPIService } from '../core/selection';
import { DisplayDensityToken } from '../core/density';
/**
 * @hidden \@internal
 */
var IgxSelectToggleIconDirective = /** @class */ (function () {
    function IgxSelectToggleIconDirective() {
    }
    IgxSelectToggleIconDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxSelectToggleIcon]'
                },] }
    ];
    return IgxSelectToggleIconDirective;
}());
export { IgxSelectToggleIconDirective };
/** @type {?} */
var noop = function () { };
var ɵ0 = noop;
/**
 * **Ignite UI for Angular Select** -
 * [Documentation](https://www.infragistics.com/products/ignite-ui-angular/angular/components/select.html)
 *
 * The `igxSelect` provides an input with dropdown list allowing selection of a single item.
 *
 * Example:
 * ```html
 * <igx-select #select1 [placeholder]="'Pick One'">
 *   <label igxLabel>Select Label</label>
 *   <igx-select-item *ngFor="let item of items" [value]="item.field">
 *     {{ item.field }}
 *   </igx-select-item>
 * </igx-select>
 * ```
 */
var IgxSelectComponent = /** @class */ (function (_super) {
    tslib_1.__extends(IgxSelectComponent, _super);
    function IgxSelectComponent(elementRef, cdr, selection, _displayDensityOptions, _injector) {
        var _this = _super.call(this, elementRef, cdr, selection, _displayDensityOptions) || this;
        _this.elementRef = elementRef;
        _this.cdr = cdr;
        _this.selection = selection;
        _this._displayDensityOptions = _displayDensityOptions;
        _this._injector = _injector;
        _this.ngControl = null;
        /**
         * @hidden \@internal do not use the drop-down container class
         */
        _this.cssClass = false;
        /**
         * @hidden \@internal
         */
        _this.allowItemsFocus = false;
        /**
         * An \@Input property that disables the `IgxSelectComponent`.
         * ```html
         * <igx-select [disabled]="'true'"></igx-select>
         * ```
         */
        _this.disabled = false;
        /**
         * @hidden \@internal
         */
        _this.maxHeight = '256px';
        /**
         * An \@Input property that sets how the select will be styled.
         * The allowed values are `line`, `box` and `border`. The default is `line`.
         * ```html
         * <igx-select [type]="'box'"></igx-select>
         * ```
         */
        _this.type = 'line';
        /**
         * The custom template, if any, that should be used when rendering the select TOGGLE(open/close) button
         *
         * ```typescript
         * // Set in typescript
         * const myCustomTemplate: TemplateRef<any> = myComponent.customTemplate;
         * myComponent.select.toggleIconTemplate = myCustomTemplate;
         * ```
         * ```html
         * <!-- Set in markup -->
         *  <igx-select #select>
         *      ...
         *      <ng-template igxSelectToggleIcon let-collapsed>
         *          <igx-icon>{{ collapsed ? 'remove_circle' : 'remove_circle_outline'}}</igx-icon>
         *      </ng-template>
         *  </igx-select>
         * ```
         */
        _this.toggleIconTemplate = null;
        //#region ControlValueAccessor
        /**
         * @hidden \@internal
         */
        _this._onChangeCallback = noop;
        /**
         * @hidden \@internal
         */
        _this._onTouchedCallback = noop;
        /**
         * @hidden \@internal
         */
        _this.writeValue = function (value) {
            _this.value = value;
        };
        return _this;
    }
    Object.defineProperty(IgxSelectComponent.prototype, "value", {
        /**
         * An @Input property that gets/sets the component value.
         *
         * ```typescript
         * // get
         * let selectValue = this.select.value;
         * ```
         *
         * ```typescript
         * // set
         * this.select.value = 'London';
         * ```
         * ```html
         * <igx-select [value]="value"></igx-select>
         * ```
         */
        get: /**
         * An \@Input property that gets/sets the component value.
         *
         * ```typescript
         * // get
         * let selectValue = this.select.value;
         * ```
         *
         * ```typescript
         * // set
         * this.select.value = 'London';
         * ```
         * ```html
         * <igx-select [value]="value"></igx-select>
         * ```
         * @return {?}
         */
        function () {
            return this._value;
        },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            var _this = this;
            if (this._value === v) {
                return;
            }
            this._value = v;
            this.setSelection(this.items.find(function (x) { return x.value === _this.value; }));
            this.cdr.detectChanges();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxSelectComponent.prototype, "selectionValue", {
        /** @hidden @internal */
        get: /**
         * @hidden \@internal
         * @return {?}
         */
        function () {
            /** @type {?} */
            var selectedItem = this.selectedItem;
            return selectedItem ? selectedItem.itemText : '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxSelectComponent.prototype, "selectedItem", {
        /** @hidden @internal */
        get: /**
         * @hidden \@internal
         * @return {?}
         */
        function () {
            return this.selection.first_item(this.id);
        },
        enumerable: true,
        configurable: true
    });
    /** @hidden @internal */
    /**
     * @hidden \@internal
     * @param {?} fn
     * @return {?}
     */
    IgxSelectComponent.prototype.registerOnChange = /**
     * @hidden \@internal
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._onChangeCallback = fn;
    };
    /** @hidden @internal */
    /**
     * @hidden \@internal
     * @param {?} fn
     * @return {?}
     */
    IgxSelectComponent.prototype.registerOnTouched = /**
     * @hidden \@internal
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._onTouchedCallback = fn;
    };
    /** @hidden @internal */
    /**
     * @hidden \@internal
     * @param {?} isDisabled
     * @return {?}
     */
    IgxSelectComponent.prototype.setDisabledState = /**
     * @hidden \@internal
     * @param {?} isDisabled
     * @return {?}
     */
    function (isDisabled) {
        this.disabled = isDisabled;
    };
    //#endregion
    /** @hidden @internal */
    //#endregion
    /**
     * @hidden \@internal
     * @return {?}
     */
    IgxSelectComponent.prototype.getEditElement = 
    //#endregion
    /**
     * @hidden \@internal
     * @return {?}
     */
    function () {
        return this.input.nativeElement;
    };
    /** @hidden @internal */
    /**
     * @hidden \@internal
     * @param {?} newSelection
     * @param {?=} event
     * @return {?}
     */
    IgxSelectComponent.prototype.selectItem = /**
     * @hidden \@internal
     * @param {?} newSelection
     * @param {?=} event
     * @return {?}
     */
    function (newSelection, event) {
        /** @type {?} */
        var oldSelection = this.selectedItem;
        if (event) {
            this.toggleDirective.close();
        }
        if (newSelection === null || newSelection === oldSelection || newSelection.disabled || newSelection.isHeader) {
            return;
        }
        /** @type {?} */
        var args = { oldSelection: oldSelection, newSelection: newSelection, cancel: false };
        this.onSelection.emit(args);
        if (args.cancel) {
            return;
        }
        this.setSelection(newSelection);
        this._value = newSelection.value;
        this.cdr.detectChanges();
        this._onChangeCallback(this.value);
    };
    /** @hidden @internal */
    /**
     * @hidden \@internal
     * @return {?}
     */
    IgxSelectComponent.prototype.getFirstItemElement = /**
     * @hidden \@internal
     * @return {?}
     */
    function () {
        return this.children.first.element.nativeElement;
    };
    /**
     * Opens the select
     *
     * ```typescript
     * this.select.open();
     * ```
     */
    /**
     * Opens the select
     *
     * ```typescript
     * this.select.open();
     * ```
     * @param {?=} overlaySettings
     * @return {?}
     */
    IgxSelectComponent.prototype.open = /**
     * Opens the select
     *
     * ```typescript
     * this.select.open();
     * ```
     * @param {?=} overlaySettings
     * @return {?}
     */
    function (overlaySettings) {
        if (this.disabled || this.items.length === 0) {
            return;
        }
        if (!this.selectedItem) {
            this.navigateFirst();
        }
        _super.prototype.open.call(this, Object.assign({}, this._overlayDefaults, this.overlaySettings, overlaySettings));
    };
    /** @hidden @internal */
    /**
     * @hidden \@internal
     * @return {?}
     */
    IgxSelectComponent.prototype.ngAfterContentInit = /**
     * @hidden \@internal
     * @return {?}
     */
    function () {
        var _this = this;
        this._overlayDefaults = {
            modal: false,
            closeOnOutsideClick: false,
            positionStrategy: new SelectPositioningStrategy(this, { target: this.inputGroup.element.nativeElement }),
            scrollStrategy: new AbsoluteScrollStrategy(),
            excludePositionTarget: true
        };
        this.children.changes.subscribe(function () {
            _this.setSelection(_this.items.find(function (x) { return x.value === _this.value; }));
            _this.cdr.detectChanges();
        });
        Promise.resolve().then(function () { return _this.children.notifyOnChanges(); });
    };
    /** @hidden @internal */
    /**
     * @hidden \@internal
     * @param {?} event
     * @return {?}
     */
    IgxSelectComponent.prototype.onToggleOpening = /**
     * @hidden \@internal
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.onOpening.emit(event);
        if (event.cancel) {
            return;
        }
        this.scrollToItem(this.selectedItem);
    };
    /**
     * @protected
     * @param {?} direction
     * @param {?=} currentIndex
     * @return {?}
     */
    IgxSelectComponent.prototype.navigate = /**
     * @protected
     * @param {?} direction
     * @param {?=} currentIndex
     * @return {?}
     */
    function (direction, currentIndex) {
        if (this.collapsed && this.selectedItem) {
            this.navigateItem(this.selectedItem.itemIndex);
        }
        _super.prototype.navigate.call(this, direction, currentIndex);
    };
    /**
     * @protected
     * @return {?}
     */
    IgxSelectComponent.prototype.manageRequiredAsterisk = /**
     * @protected
     * @return {?}
     */
    function () {
        if (this.ngControl && this.ngControl.control.validator) {
            // Run the validation with empty object to check if required is enabled.
            /** @type {?} */
            var error = this.ngControl.control.validator((/** @type {?} */ ({})));
            this.inputGroup.isRequired = error && error.required;
            this.cdr.markForCheck();
        }
    };
    /**
     * @private
     * @param {?} item
     * @return {?}
     */
    IgxSelectComponent.prototype.setSelection = /**
     * @private
     * @param {?} item
     * @return {?}
     */
    function (item) {
        if (item && item.value !== undefined && item.value !== null) {
            this.selection.set(this.id, new Set([item]));
        }
        else {
            this.selection.clear(this.id);
        }
    };
    /** @hidden @internal */
    /**
     * @hidden \@internal
     * @return {?}
     */
    IgxSelectComponent.prototype.onBlur = /**
     * @hidden \@internal
     * @return {?}
     */
    function () {
        this._onTouchedCallback();
        if (this.ngControl && !this.ngControl.valid) {
            this.input.valid = IgxInputState.INVALID;
        }
        else {
            this.input.valid = IgxInputState.INITIAL;
        }
        if (!this.collapsed) {
            this.toggleDirective.close();
        }
    };
    /** @hidden @internal */
    /**
     * @hidden \@internal
     * @return {?}
     */
    IgxSelectComponent.prototype.onFocus = /**
     * @hidden \@internal
     * @return {?}
     */
    function () {
        this._onTouchedCallback();
    };
    /**
     * @protected
     * @return {?}
     */
    IgxSelectComponent.prototype.onStatusChanged = /**
     * @protected
     * @return {?}
     */
    function () {
        if ((this.ngControl.control.touched || this.ngControl.control.dirty) &&
            (this.ngControl.control.validator || this.ngControl.control.asyncValidator)) {
            if (this.inputGroup.isFocused) {
                this.input.valid = this.ngControl.valid ? IgxInputState.VALID : IgxInputState.INVALID;
            }
            else {
                this.input.valid = this.ngControl.valid ? IgxInputState.INITIAL : IgxInputState.INVALID;
            }
        }
        this.manageRequiredAsterisk();
    };
    /**
     * @hidden @internal
     */
    /**
     * @hidden \@internal
     * @return {?}
     */
    IgxSelectComponent.prototype.ngOnInit = /**
     * @hidden \@internal
     * @return {?}
     */
    function () {
        this.ngControl = this._injector.get((/** @type {?} */ (NgControl)), null);
    };
    /**
     * @hidden @internal
     */
    /**
     * @hidden \@internal
     * @return {?}
     */
    IgxSelectComponent.prototype.ngAfterViewInit = /**
     * @hidden \@internal
     * @return {?}
     */
    function () {
        if (this.ngControl) {
            this._statusChanges$ = this.ngControl.statusChanges.subscribe(this.onStatusChanged.bind(this));
            this.manageRequiredAsterisk();
        }
        this.cdr.detectChanges();
    };
    /**
     * @hidden @internal
     */
    /**
     * @hidden \@internal
     * @return {?}
     */
    IgxSelectComponent.prototype.ngOnDestroy = /**
     * @hidden \@internal
     * @return {?}
     */
    function () {
        this.selection.clear(this.id);
        if (this._statusChanges$) {
            this._statusChanges$.unsubscribe();
        }
    };
    IgxSelectComponent.decorators = [
        { type: Component, args: [{
                    selector: 'igx-select',
                    template: "<igx-input-group #inputGroup class=\"input-group\" [disabled]=\"disabled\" (click)=\"toggle()\" [type]=\"type\" [displayDensity]=\"displayDensity\">\n    <ng-container ngProjectAs=\"[igxLabel]\">\n        <ng-content select=\"[igxLabel]\"></ng-content>\n    </ng-container>\n    <ng-container ngProjectAs=\"igx-prefix\">\n        <ng-content select=\"igx-prefix,[igxPrefix]\"></ng-content>\n    </ng-container>\n        <input #input class=\"input\" type=\"text\" igxInput [igxSelectItemNavigation]=\"this\"\n            readonly=\"true\"\n            [attr.placeholder]=\"this.placeholder\"\n            [value]=\"this.selectionValue\"\n            role=\"combobox\"\n            aria-haspopup=\"listbox\"\n            [attr.aria-labelledby]=\"this.label ? this.label.id : ''\"\n            [attr.aria-expanded]=\"!this.collapsed\"\n            [attr.aria-owns]=\"this.listId\"\n            [attr.aria-activedescendant]=\"!this.collapsed ? this.focusedItem?.id : null\"\n            (blur)=\"onBlur()\"\n            (focus)=\"onFocus()\"\n            />\n        <ng-container ngProjectAs=\"igx-suffix\">\n                <ng-content select=\"igx-suffix,[igxSuffix]\"></ng-content>\n            </ng-container>\n        <igx-suffix>\n            <ng-container *ngIf=\"toggleIconTemplate\">\n                <ng-container *ngTemplateOutlet=\"toggleIconTemplate; context: {$implicit: this.collapsed}\"></ng-container>\n                </ng-container>\n            <igx-icon *ngIf=\"!toggleIconTemplate\" fontSet=\"material\">{{ collapsed ? 'arrow_drop_down' : 'arrow_drop_up'}}</igx-icon>\n        </igx-suffix>\n    </igx-input-group>\n    <div class=\"igx-drop-down__list--select\" igxToggle [style.maxHeight]=\"maxHeight\"\n        (onOpening)=\"onToggleOpening($event)\" (onOpened)=\"onToggleOpened()\"\n        (onClosing)=\"onToggleClosing($event)\" (onClosed)=\"onToggleClosed()\"\n        [attr.id]=\"this.listId\" role=\"listbox\">\n        <ng-content select=\"igx-select-item, igx-select-item-group\">\n        </ng-content>\n    </div>\n",
                    providers: [
                        { provide: NG_VALUE_ACCESSOR, useExisting: IgxSelectComponent, multi: true },
                        { provide: IGX_DROPDOWN_BASE, useExisting: IgxSelectComponent }
                    ],
                    styles: ["\n        :host {\n            display: block;\n        }\n    "]
                }] }
    ];
    /** @nocollapse */
    IgxSelectComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ChangeDetectorRef },
        { type: IgxSelectionAPIService },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [DisplayDensityToken,] }] },
        { type: Injector }
    ]; };
    IgxSelectComponent.propDecorators = {
        inputGroup: [{ type: ViewChild, args: ['inputGroup', { read: IgxInputGroupComponent, static: true },] }],
        input: [{ type: ViewChild, args: ['input', { read: IgxInputDirective, static: true },] }],
        children: [{ type: ContentChildren, args: [forwardRef(function () { return IgxSelectItemComponent; }), { descendants: true },] }],
        label: [{ type: ContentChild, args: [forwardRef(function () { return IgxLabelDirective; }), { static: true },] }],
        value: [{ type: Input }],
        placeholder: [{ type: Input }],
        disabled: [{ type: Input }],
        overlaySettings: [{ type: Input }],
        maxHeight: [{ type: HostBinding, args: ['style.maxHeight',] }],
        type: [{ type: Input }],
        toggleIconTemplate: [{ type: ContentChild, args: [IgxSelectToggleIconDirective, { read: TemplateRef, static: false },] }]
    };
    return IgxSelectComponent;
}(IgxDropDownComponent));
export { IgxSelectComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    IgxSelectComponent.prototype.ngControl;
    /**
     * @type {?}
     * @private
     */
    IgxSelectComponent.prototype._statusChanges$;
    /**
     * @type {?}
     * @private
     */
    IgxSelectComponent.prototype._overlayDefaults;
    /**
     * @type {?}
     * @private
     */
    IgxSelectComponent.prototype._value;
    /**
     * @hidden \@internal do not use the drop-down container class
     * @type {?}
     */
    IgxSelectComponent.prototype.cssClass;
    /**
     * @hidden \@internal
     * @type {?}
     */
    IgxSelectComponent.prototype.inputGroup;
    /**
     * @hidden \@internal
     * @type {?}
     */
    IgxSelectComponent.prototype.input;
    /**
     * @hidden \@internal
     * @type {?}
     */
    IgxSelectComponent.prototype.children;
    /**
     * @hidden \@internal
     * @type {?}
     */
    IgxSelectComponent.prototype.label;
    /**
     * @hidden \@internal
     * @type {?}
     */
    IgxSelectComponent.prototype.allowItemsFocus;
    /**
     * @hidden \@internal
     * @type {?}
     */
    IgxSelectComponent.prototype.height;
    /**
     * An \@Input property that sets input placeholder.
     *
     * @type {?}
     */
    IgxSelectComponent.prototype.placeholder;
    /**
     * An \@Input property that disables the `IgxSelectComponent`.
     * ```html
     * <igx-select [disabled]="'true'"></igx-select>
     * ```
     * @type {?}
     */
    IgxSelectComponent.prototype.disabled;
    /**
     * An \@Input property that sets custom OverlaySettings `IgxSelectComponent`.
     * ```html
     * <igx-select [overlaySettings] = "customOverlaySettings"></igx-select>
     * ```
     * @type {?}
     */
    IgxSelectComponent.prototype.overlaySettings;
    /**
     * @hidden \@internal
     * @type {?}
     */
    IgxSelectComponent.prototype.maxHeight;
    /**
     * @hidden \@internal
     * @type {?}
     */
    IgxSelectComponent.prototype.width;
    /**
     * An \@Input property that sets how the select will be styled.
     * The allowed values are `line`, `box` and `border`. The default is `line`.
     * ```html
     * <igx-select [type]="'box'"></igx-select>
     * ```
     * @type {?}
     */
    IgxSelectComponent.prototype.type;
    /**
     * The custom template, if any, that should be used when rendering the select TOGGLE(open/close) button
     *
     * ```typescript
     * // Set in typescript
     * const myCustomTemplate: TemplateRef<any> = myComponent.customTemplate;
     * myComponent.select.toggleIconTemplate = myCustomTemplate;
     * ```
     * ```html
     * <!-- Set in markup -->
     *  <igx-select #select>
     *      ...
     *      <ng-template igxSelectToggleIcon let-collapsed>
     *          <igx-icon>{{ collapsed ? 'remove_circle' : 'remove_circle_outline'}}</igx-icon>
     *      </ng-template>
     *  </igx-select>
     * ```
     * @type {?}
     */
    IgxSelectComponent.prototype.toggleIconTemplate;
    /**
     * @hidden \@internal
     * @type {?}
     * @private
     */
    IgxSelectComponent.prototype._onChangeCallback;
    /**
     * @hidden \@internal
     * @type {?}
     * @private
     */
    IgxSelectComponent.prototype._onTouchedCallback;
    /**
     * @hidden \@internal
     * @type {?}
     */
    IgxSelectComponent.prototype.writeValue;
    /**
     * @type {?}
     * @protected
     */
    IgxSelectComponent.prototype.elementRef;
    /**
     * @type {?}
     * @protected
     */
    IgxSelectComponent.prototype.cdr;
    /**
     * @type {?}
     * @protected
     */
    IgxSelectComponent.prototype.selection;
    /**
     * @type {?}
     * @protected
     */
    IgxSelectComponent.prototype._displayDensityOptions;
    /**
     * @type {?}
     * @private
     */
    IgxSelectComponent.prototype._injector;
}
export { ɵ0 };
