// igniteui-angular/esm2015/lib/directives/for-of/horizontal.virtual.helper.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, HostBinding, Input, ViewChild, ViewContainerRef, ChangeDetectorRef } from '@angular/core';
import { VirtualHelperBaseDirective } from './base.helper.component';
/**
 * @hidden
 */
export class HVirtualHelperComponent extends VirtualHelperBaseDirective {
    /**
     * @param {?} elementRef
     * @param {?} cdr
     */
    constructor(elementRef, cdr) {
        super(elementRef, cdr);
        this.elementRef = elementRef;
        this.cdr = cdr;
        this.cssClasses = 'igx-vhelper--horizontal';
    }
}
HVirtualHelperComponent.decorators = [
    { type: Component, args: [{
                selector: 'igx-horizontal-virtual-helper',
                template: '<div #horizontal_container class="igx-vhelper__placeholder-content" [style.width.px]="size"></div>'
            }] }
];
/** @nocollapse */
HVirtualHelperComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: ChangeDetectorRef }
];
HVirtualHelperComponent.propDecorators = {
    _vcr: [{ type: ViewChild, args: ['horizontal_container', { read: ViewContainerRef, static: true },] }],
    width: [{ type: Input }],
    cssClasses: [{ type: HostBinding, args: ['class',] }]
};
if (false) {
    /** @type {?} */
    HVirtualHelperComponent.prototype._vcr;
    /** @type {?} */
    HVirtualHelperComponent.prototype.width;
    /** @type {?} */
    HVirtualHelperComponent.prototype.cssClasses;
    /** @type {?} */
    HVirtualHelperComponent.prototype.elementRef;
    /** @type {?} */
    HVirtualHelperComponent.prototype.cdr;
}
