// igniteui-angular/esm2015/lib/grids/drag-select.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Output, EventEmitter, ElementRef, NgZone } from '@angular/core';
import { interval, Subject } from 'rxjs';
import { filter, takeUntil } from 'rxjs/operators';
/** @enum {number} */
const DragScrollDirection = {
    NONE: 0,
    LEFT: 1,
    TOP: 2,
    RIGHT: 3,
    BOTTOM: 4,
    TOPLEFT: 5,
    TOPRIGHT: 6,
    BOTTOMLEFT: 7,
    BOTTOMRIGHT: 8,
};
export { DragScrollDirection };
DragScrollDirection[DragScrollDirection.NONE] = 'NONE';
DragScrollDirection[DragScrollDirection.LEFT] = 'LEFT';
DragScrollDirection[DragScrollDirection.TOP] = 'TOP';
DragScrollDirection[DragScrollDirection.RIGHT] = 'RIGHT';
DragScrollDirection[DragScrollDirection.BOTTOM] = 'BOTTOM';
DragScrollDirection[DragScrollDirection.TOPLEFT] = 'TOPLEFT';
DragScrollDirection[DragScrollDirection.TOPRIGHT] = 'TOPRIGHT';
DragScrollDirection[DragScrollDirection.BOTTOMLEFT] = 'BOTTOMLEFT';
DragScrollDirection[DragScrollDirection.BOTTOMRIGHT] = 'BOTTOMRIGHT';
export class IgxGridDragSelectDirective {
    /**
     * @param {?} ref
     * @param {?} zone
     */
    constructor(ref, zone) {
        this.ref = ref;
        this.zone = zone;
        this.onDragStop = new EventEmitter();
        this.onDragScroll = new EventEmitter();
        this.end$ = new Subject();
        this.lastDirection = DragScrollDirection.NONE;
        this.startDragSelection = (ev) => {
            if (!this.activeDrag) {
                return;
            }
            /** @type {?} */
            const x = ev.clientX;
            /** @type {?} */
            const y = ev.clientY;
            /** @type {?} */
            const direction = this._measureDimensions(x, y);
            if (direction === this.lastDirection) {
                return;
            }
            this.unsubscribe();
            this._sub = this._interval$.subscribe(() => this.onDragScroll.emit(direction));
            this.lastDirection = direction;
        };
        this.stopDragSelection = () => {
            if (!this.activeDrag) {
                return;
            }
            this.onDragStop.emit(false);
            this.unsubscribe();
            this.lastDirection = DragScrollDirection.NONE;
        };
        this._interval$ = interval(100).pipe(takeUntil(this.end$), filter(() => this.activeDrag));
    }
    /**
     * @return {?}
     */
    get activeDrag() {
        return this._activeDrag;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set activeDrag(val) {
        if (val !== this._activeDrag) {
            this.unsubscribe();
            this._activeDrag = val;
        }
    }
    /**
     * @return {?}
     */
    get nativeElement() {
        return this.ref.nativeElement;
    }
    /**
     * @return {?}
     */
    get clientRect() {
        return this.nativeElement.getBoundingClientRect();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.zone.runOutsideAngular(() => {
            this.nativeElement.addEventListener('pointerover', this.startDragSelection);
            this.nativeElement.addEventListener('pointerleave', this.stopDragSelection);
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.zone.runOutsideAngular(() => {
            this.nativeElement.removeEventListener('pointerover', this.startDragSelection);
            this.nativeElement.removeEventListener('pointerleave', this.stopDragSelection);
        });
        this.unsubscribe();
        this.end$.complete();
    }
    /**
     * @param {?} x
     * @param {?} y
     * @return {?}
     */
    _measureDimensions(x, y) {
        /** @type {?} */
        let direction;
        /** @type {?} */
        const rect = this.clientRect;
        /** @type {?} */
        const RATIO = 0.15;
        /** @type {?} */
        const offsetX = Math.trunc(x - rect.left);
        /** @type {?} */
        const offsetY = Math.trunc(y - rect.top);
        /** @type {?} */
        const left = offsetX <= rect.width * RATIO;
        /** @type {?} */
        const right = offsetX >= rect.width * (1 - RATIO);
        /** @type {?} */
        const top = offsetY <= rect.height * RATIO;
        /** @type {?} */
        const bottom = offsetY >= rect.height * (1 - RATIO);
        if (top && left) {
            direction = DragScrollDirection.TOPLEFT;
        }
        else if (top && right) {
            direction = DragScrollDirection.TOPRIGHT;
        }
        else if (bottom && left) {
            direction = DragScrollDirection.BOTTOMLEFT;
        }
        else if (bottom && right) {
            direction = DragScrollDirection.BOTTOMRIGHT;
        }
        else if (top) {
            direction = DragScrollDirection.TOP;
        }
        else if (bottom) {
            direction = DragScrollDirection.BOTTOM;
        }
        else if (left) {
            direction = DragScrollDirection.LEFT;
        }
        else if (right) {
            direction = DragScrollDirection.RIGHT;
        }
        else {
            direction = DragScrollDirection.NONE;
        }
        return direction;
    }
    /**
     * @protected
     * @return {?}
     */
    unsubscribe() {
        if (this._sub) {
            this._sub.unsubscribe();
        }
    }
}
IgxGridDragSelectDirective.decorators = [
    { type: Directive, args: [{
                selector: '[igxGridDragSelect]'
            },] }
];
/** @nocollapse */
IgxGridDragSelectDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: NgZone }
];
IgxGridDragSelectDirective.propDecorators = {
    activeDrag: [{ type: Input, args: ['igxGridDragSelect',] }],
    onDragStop: [{ type: Output }],
    onDragScroll: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    IgxGridDragSelectDirective.prototype._activeDrag;
    /** @type {?} */
    IgxGridDragSelectDirective.prototype.onDragStop;
    /** @type {?} */
    IgxGridDragSelectDirective.prototype.onDragScroll;
    /**
     * @type {?}
     * @protected
     */
    IgxGridDragSelectDirective.prototype.end$;
    /**
     * @type {?}
     * @protected
     */
    IgxGridDragSelectDirective.prototype.lastDirection;
    /**
     * @type {?}
     * @protected
     */
    IgxGridDragSelectDirective.prototype._interval$;
    /**
     * @type {?}
     * @protected
     */
    IgxGridDragSelectDirective.prototype._sub;
    /** @type {?} */
    IgxGridDragSelectDirective.prototype.startDragSelection;
    /** @type {?} */
    IgxGridDragSelectDirective.prototype.stopDragSelection;
    /**
     * @type {?}
     * @private
     */
    IgxGridDragSelectDirective.prototype.ref;
    /**
     * @type {?}
     * @private
     */
    IgxGridDragSelectDirective.prototype.zone;
}
