// igniteui-angular/esm2015/lib/grids/igx-row-selectors.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, NgModule, TemplateRef } from '@angular/core';
/**
 * @hidden
 */
export class IgxRowSelectorDirective {
    /**
     * @param {?} templateRef
     */
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
}
IgxRowSelectorDirective.decorators = [
    { type: Directive, args: [{
                selector: '[igxRowSelector]'
            },] }
];
/** @nocollapse */
IgxRowSelectorDirective.ctorParameters = () => [
    { type: TemplateRef }
];
if (false) {
    /** @type {?} */
    IgxRowSelectorDirective.prototype.templateRef;
}
/**
 * @hidden
 */
export class IgxHeadSelectorDirective {
    /**
     * @param {?} templateRef
     */
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
}
IgxHeadSelectorDirective.decorators = [
    { type: Directive, args: [{
                selector: '[igxHeadSelector]'
            },] }
];
/** @nocollapse */
IgxHeadSelectorDirective.ctorParameters = () => [
    { type: TemplateRef }
];
if (false) {
    /** @type {?} */
    IgxHeadSelectorDirective.prototype.templateRef;
}
/**
 * @hidden
 */
export class IgxRowSelectorsModule {
}
IgxRowSelectorsModule.decorators = [
    { type: NgModule, args: [{
                declarations: [IgxRowSelectorDirective, IgxHeadSelectorDirective],
                exports: [IgxRowSelectorDirective, IgxHeadSelectorDirective]
            },] }
];
