// igniteui-angular/esm5/lib/grids/filtering/excel-style/excel-style-search.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectionStrategy, Input, ViewChild, ChangeDetectorRef, TemplateRef, Directive } from '@angular/core';
import { IgxColumnComponent } from '../../column.component';
import { IgxInputDirective } from '../../../directives/input/input.directive';
import { DisplayDensity } from '../../../core/density';
import { IgxForOfDirective } from '../../../directives/for-of/for_of.directive';
var IgxExcelStyleLoadingValuesTemplateDirective = /** @class */ (function () {
    function IgxExcelStyleLoadingValuesTemplateDirective(template) {
        this.template = template;
    }
    IgxExcelStyleLoadingValuesTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[igxExcelStyleLoading]'
                },] }
    ];
    /** @nocollapse */
    IgxExcelStyleLoadingValuesTemplateDirective.ctorParameters = function () { return [
        { type: TemplateRef }
    ]; };
    return IgxExcelStyleLoadingValuesTemplateDirective;
}());
export { IgxExcelStyleLoadingValuesTemplateDirective };
if (false) {
    /** @type {?} */
    IgxExcelStyleLoadingValuesTemplateDirective.prototype.template;
}
/**
 * @hidden
 */
var IgxExcelStyleSearchComponent = /** @class */ (function () {
    function IgxExcelStyleSearchComponent(cdr) {
        this.cdr = cdr;
    }
    Object.defineProperty(IgxExcelStyleSearchComponent.prototype, "isLoading", {
        get: /**
         * @return {?}
         */
        function () {
            return this._isLoading;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._isLoading = value;
            if (!((/** @type {?} */ (this.cdr))).destroyed) {
                this.cdr.detectChanges();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxExcelStyleSearchComponent.prototype, "valuesLoadingTemplate", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.grid.excelStyleLoadingValuesTemplateDirective) {
                return this.grid.excelStyleLoadingValuesTemplateDirective.template;
            }
            else {
                return this.defaultExcelStyleLoadingValuesTemplate;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    IgxExcelStyleSearchComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.refreshSize();
    };
    /**
     * @return {?}
     */
    IgxExcelStyleSearchComponent.prototype.refreshSize = /**
     * @return {?}
     */
    function () {
        var _this = this;
        requestAnimationFrame(function () {
            _this.virtDir.recalcUpdateSizes();
        });
    };
    /**
     * @return {?}
     */
    IgxExcelStyleSearchComponent.prototype.clearInput = /**
     * @return {?}
     */
    function () {
        this.searchValue = null;
    };
    /**
     * @param {?} eventArgs
     * @return {?}
     */
    IgxExcelStyleSearchComponent.prototype.onCheckboxChange = /**
     * @param {?} eventArgs
     * @return {?}
     */
    function (eventArgs) {
        var _this = this;
        /** @type {?} */
        var selectedIndex = this.data.indexOf(eventArgs.checkbox.value);
        if (selectedIndex === 0) {
            this.data.forEach(function (element) {
                element.isSelected = eventArgs.checked;
                _this.data[0].indeterminate = false;
            });
        }
        else {
            eventArgs.checkbox.value.isSelected = eventArgs.checked;
            if (!this.data.slice(1, this.data.length).find(function (el) { return el.isSelected === false; })) {
                this.data[0].indeterminate = false;
                this.data[0].isSelected = true;
            }
            else if (!this.data.slice(1, this.data.length).find(function (el) { return el.isSelected === true; })) {
                this.data[0].indeterminate = false;
                this.data[0].isSelected = false;
            }
            else {
                this.data[0].indeterminate = true;
            }
        }
        eventArgs.checkbox.nativeCheckbox.nativeElement.blur();
    };
    Object.defineProperty(IgxExcelStyleSearchComponent.prototype, "itemSize", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var itemSize = '40px';
            switch (this.displayDensity) {
                case DisplayDensity.cosy:
                    itemSize = '32px';
                    break;
                case DisplayDensity.compact:
                    itemSize = '24px';
                    break;
                default: break;
            }
            return itemSize;
        },
        enumerable: true,
        configurable: true
    });
    IgxExcelStyleSearchComponent.decorators = [
        { type: Component, args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    preserveWhitespaces: false,
                    selector: 'igx-excel-style-search',
                    template: "<igx-input-group\n            type=\"box\"\n            [displayDensity]=\"displayDensity\"\n            [supressInputAutofocus]=\"true\">\n    <igx-icon igxPrefix>search</igx-icon>\n    <input\n        #input\n        igxInput\n        tabindex=\"0\"\n        [(ngModel)]=\"searchValue\"\n        [placeholder]=\"column.grid.resourceStrings.igx_grid_excel_search_placeholder\"\n        autocomplete=\"off\"/>\n    <igx-icon\n        igxSuffix\n        *ngIf=\"searchValue || searchValue === 0\"\n        (click)=\"clearInput()\"\n        tabindex=\"0\">\n        clear\n    </igx-icon>\n</igx-input-group>\n\n<igx-list [displayDensity]=\"displayDensity\" [style.height.px]=\"250\" [isLoading]=\"isLoading\">\n    <div [style.overflow]=\"'hidden'\" [style.position]=\"'relative'\">\n        <igx-list-item\n            *igxFor=\"let item of data | excelStyleSearchFilter: searchValue; scrollOrientation : 'vertical'; containerSize: '250px'; itemSize: itemSize\">\n            <igx-checkbox\n            [value]=\"item\"\n            tabindex=\"-1\"\n            [checked]=\"item.isSelected\"\n            [disableRipple]=\"true\"\n            [indeterminate]=\"item.indeterminate\"\n            [disableTransitions]=\"true\"\n            (change)=\"onCheckboxChange($event)\">\n                {{ column.formatter && !item.isSpecial ? column.formatter(item.label) : column.dataType === 'number' ? (item.label | igxdecimal:\n                    column.grid.locale) : column.dataType === 'date' ? (item.label | igxdate: column.grid.locale) : item.label }}\n            </igx-checkbox>\n        </igx-list-item>\n    </div>\n\n    <ng-template igxDataLoading>        \n        <div class=\"igx-excel-filter__loading\">\n            <ng-container *ngTemplateOutlet=\"valuesLoadingTemplate\">\n            </ng-container>\n        </div>\n    </ng-template>\n</igx-list>\n\n<ng-template #defaultExcelStyleLoadingValuesTemplate>\n    <igx-circular-bar [indeterminate]=\"true\">\n    </igx-circular-bar>\n</ng-template>\n"
                }] }
    ];
    /** @nocollapse */
    IgxExcelStyleSearchComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    IgxExcelStyleSearchComponent.propDecorators = {
        grid: [{ type: Input }],
        data: [{ type: Input }],
        column: [{ type: Input }],
        searchInput: [{ type: ViewChild, args: ['input', { read: IgxInputDirective, static: true },] }],
        displayDensity: [{ type: Input }],
        virtDir: [{ type: ViewChild, args: [IgxForOfDirective, { static: true },] }],
        defaultExcelStyleLoadingValuesTemplate: [{ type: ViewChild, args: ['defaultExcelStyleLoadingValuesTemplate', { read: TemplateRef, static: true },] }]
    };
    return IgxExcelStyleSearchComponent;
}());
export { IgxExcelStyleSearchComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    IgxExcelStyleSearchComponent.prototype._isLoading;
    /** @type {?} */
    IgxExcelStyleSearchComponent.prototype.searchValue;
    /** @type {?} */
    IgxExcelStyleSearchComponent.prototype.grid;
    /** @type {?} */
    IgxExcelStyleSearchComponent.prototype.data;
    /** @type {?} */
    IgxExcelStyleSearchComponent.prototype.column;
    /** @type {?} */
    IgxExcelStyleSearchComponent.prototype.searchInput;
    /** @type {?} */
    IgxExcelStyleSearchComponent.prototype.displayDensity;
    /**
     * @type {?}
     * @protected
     */
    IgxExcelStyleSearchComponent.prototype.virtDir;
    /**
     * @type {?}
     * @protected
     */
    IgxExcelStyleSearchComponent.prototype.defaultExcelStyleLoadingValuesTemplate;
    /** @type {?} */
    IgxExcelStyleSearchComponent.prototype.cdr;
}
