// igniteui-angular/esm2015/lib/combo/combo.common.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const IGX_COMBO_COMPONENT = 'IgxComboComponentToken';
/**
 * @hidden \@internal TODO: Evaluate
 * @record
 */
export function IgxComboBase() { }
if (false) {
    /** @type {?} */
    IgxComboBase.prototype.id;
    /** @type {?} */
    IgxComboBase.prototype.data;
    /** @type {?} */
    IgxComboBase.prototype.valueKey;
    /** @type {?} */
    IgxComboBase.prototype.groupKey;
    /** @type {?} */
    IgxComboBase.prototype.isRemote;
    /** @type {?} */
    IgxComboBase.prototype.filteredData;
    /** @type {?} */
    IgxComboBase.prototype.filteringExpressions;
    /** @type {?} */
    IgxComboBase.prototype.totalItemCount;
    /** @type {?} */
    IgxComboBase.prototype.itemsMaxHeight;
    /** @type {?} */
    IgxComboBase.prototype.itemHeight;
    /** @type {?} */
    IgxComboBase.prototype.searchValue;
    /** @type {?} */
    IgxComboBase.prototype.searchInput;
    /** @type {?} */
    IgxComboBase.prototype.comboInput;
    /** @type {?} */
    IgxComboBase.prototype.onOpened;
    /** @type {?} */
    IgxComboBase.prototype.onOpening;
    /** @type {?} */
    IgxComboBase.prototype.onClosing;
    /** @type {?} */
    IgxComboBase.prototype.onClosed;
    /**
     * @param {?=} opening
     * @return {?}
     */
    IgxComboBase.prototype.focusSearchInput = function (opening) { };
    /**
     * @return {?}
     */
    IgxComboBase.prototype.triggerCheck = function () { };
    /**
     * @return {?}
     */
    IgxComboBase.prototype.addItemToCollection = function () { };
    /**
     * @return {?}
     */
    IgxComboBase.prototype.isAddButtonVisible = function () { };
    /**
     * @param {?=} event
     * @return {?}
     */
    IgxComboBase.prototype.handleInputChange = function (event) { };
    /**
     * @param {?} itemID
     * @return {?}
     */
    IgxComboBase.prototype.isItemSelected = function (itemID) { };
    /**
     * @param {?} itemIDs
     * @param {?=} clearSelection
     * @param {?=} event
     * @return {?}
     */
    IgxComboBase.prototype.selectItems = function (itemIDs, clearSelection, event) { };
    /**
     * @param {?} itemIDs
     * @param {?=} event
     * @return {?}
     */
    IgxComboBase.prototype.deselectItems = function (itemIDs, event) { };
}
