// igniteui-cli/migrations/update-4_2_3/index.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
// tslint:disable:no-implicit-dependencies
const schematics_1 = require("@angular-devkit/schematics");
// tslint:disable-next-line:no-submodule-imports
const testing_1 = require("@angular-devkit/schematics/testing");
describe("Update 4.2.3", () => {
    let appTree;
    const schematicRunner = new testing_1.SchematicTestRunner("ig-migrate", path.join(__dirname, "../migration-collection.json"));
    const configJson = {
        defaultProject: "testProj",
        projects: {
            testProj: {
                sourceRoot: "/src"
            }
        }
    };
    beforeEach(() => {
        appTree = new testing_1.UnitTestTree(new schematics_1.EmptyTree());
        appTree.create("/angular.json", JSON.stringify(configJson));
    });
    it("should add Awesome Grid extra paginator style", done => {
        const stylesFile = "/src/app/awesome-grid/awesome-grid.component.scss";
        appTree.create(stylesFile, `@import '~igniteui-angular/lib/core/styles/themes/index';

:host ::ng-deep {

  .grid__wrapper {
    @include igx-grid($grid-sample-theme);
  }

  .igx-paginator>* {
    margin: 0 .3125rem;
    display: flex;
    align-items: center;
  }
}
`);
        schematicRunner.runSchematic("migration-04", {}, appTree);
        expect(appTree.readContent(stylesFile).replace(/\r\n/g, "\n"))
            .toEqual(`@import '~igniteui-angular/lib/core/styles/themes/index';

:host ::ng-deep {

  .grid__wrapper {
    @include igx-grid($grid-sample-theme);
  }

  .igx-paginator {
    display: flex;
    justify-content: center;
    align-items: center;
    padding: 0.3125rem 0;
  }

  .igx-paginator>* {
    margin: 0 .3125rem;
    display: flex;
    align-items: center;
  }
}
`.replace(/\r\n/g, "\n"));
        done();
    });
});
