// igniteui-cli/templates/angular/ig-ts/grid/grid-export/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const GridHelper_1 = require("../../../../../lib/project-utility/GridHelper");
const AngularTemplate_1 = require("../../../../../lib/templates/AngularTemplate");
class GridExportTemplate extends AngularTemplate_1.AngularTemplate {
    constructor() {
        super(__dirname);
        this.userExtraConfiguration = {};
        this.id = "grid-export";
        this.name = "Export Grid";
        this.controlGroup = "Data Grids";
        this.widget = "igGrid";
        this.description = "igGrid export template for Angular";
        this.dependencies = ["igGrid", "igExcel", "igGridExcelExporter"];
        this.projectType = "ig-ts";
        this.extraConfigurations = [];
        this.hasExtraConfiguration = true;
        this.listInComponentTemplates = true;
        this.gridHelper = new GridHelper_1.GridHelper();
        const featureConfiguration = {
            choices: ["Summaries", "Hiding"],
            default: "",
            key: "features",
            message: "Select features for the igGrid",
            type: 1 /* MultiChoice */
        };
        this.extraConfigurations.push(featureConfiguration);
    }
    setExtraConfiguration(extraConfigKeys) {
        this.userExtraConfiguration = extraConfigKeys;
    }
    generateFiles(projectPath, name, ...options) {
        const features = this.gridHelper.generateFeatures(this.userExtraConfiguration["features"], 3);
        const config = { "$(gridFeatures)": features };
        return super.generateFiles(projectPath, name, { extraConfig: config });
    }
    getExtraConfiguration() {
        return this.extraConfigurations;
    }
}
module.exports = new GridExportTemplate();
