// igniteui-angular/esm5/lib/combo/combo.api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @hidden
 */
var /**
 * @hidden
 */
IgxComboAPIService = /** @class */ (function () {
    function IgxComboAPIService() {
        this.disableTransitions = false;
    }
    /**
     * @param {?} combo
     * @return {?}
     */
    IgxComboAPIService.prototype.register = /**
     * @param {?} combo
     * @return {?}
     */
    function (combo) {
        this.combo = combo;
    };
    /**
     * @return {?}
     */
    IgxComboAPIService.prototype.clear = /**
     * @return {?}
     */
    function () {
        this.combo = null;
    };
    Object.defineProperty(IgxComboAPIService.prototype, "valueKey", {
        get: /**
         * @return {?}
         */
        function () {
            return this.combo.valueKey !== null && this.combo.valueKey !== undefined ? this.combo.valueKey : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxComboAPIService.prototype, "item_focusable", {
        get: /**
         * @return {?}
         */
        function () {
            return false;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxComboAPIService.prototype, "isRemote", {
        get: /**
         * @return {?}
         */
        function () {
            return this.combo.isRemote;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    IgxComboAPIService.prototype.add_custom_item = /**
     * @return {?}
     */
    function () {
        if (!this.combo) {
            return;
        }
        this.combo.addItemToCollection();
    };
    Object.defineProperty(IgxComboAPIService.prototype, "comboID", {
        get: /**
         * @return {?}
         */
        function () {
            return this.combo.id;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} itemID
     * @param {?=} event
     * @return {?}
     */
    IgxComboAPIService.prototype.set_selected_item = /**
     * @param {?} itemID
     * @param {?=} event
     * @return {?}
     */
    function (itemID, event) {
        /** @type {?} */
        var selected = this.combo.isItemSelected(itemID);
        if (itemID === null || itemID === undefined) {
            return;
        }
        if (!selected) {
            this.combo.selectItems([itemID], false, event);
        }
        else {
            this.combo.deselectItems([itemID], event);
        }
    };
    /**
     * @param {?} itemID
     * @return {?}
     */
    IgxComboAPIService.prototype.is_item_selected = /**
     * @param {?} itemID
     * @return {?}
     */
    function (itemID) {
        return this.combo.isItemSelected(itemID);
    };
    return IgxComboAPIService;
}());
/**
 * @hidden
 */
export { IgxComboAPIService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    IgxComboAPIService.prototype.combo;
    /** @type {?} */
    IgxComboAPIService.prototype.disableTransitions;
}
