// igniteui-cli/templates/angular/ig-ts/tree-grid/tree-grid-export/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const GridHelper_1 = require("../../../../../lib/project-utility/GridHelper");
const AngularTemplate_1 = require("../../../../../lib/templates/AngularTemplate");
class TreeGridExportTemplate extends AngularTemplate_1.AngularTemplate {
    constructor() {
        super(__dirname);
        this.userExtraConfiguration = {};
        this.id = "tree-grid-export";
        this.name = "Tree Grid Exporting";
        this.widget = "igTreeGrid";
        this.controlGroup = "Data Grids";
        this.description = "igTreeGrid export template for Angular";
        this.dependencies = ["igTreeGrid", "igExcel", "igGridExcelExporter"];
        this.projectType = "ig-ts";
        this.listInComponentTemplates = true;
        this.hasExtraConfiguration = true;
        this.extraConfigurations = [];
        this.gridHelper = new GridHelper_1.GridHelper();
        this.gridHelper.tree = true;
        const featureConfiguration = {
            choices: ["Filtering", "Hiding"],
            default: "",
            key: "features",
            message: "Select features for the igTreeGrid",
            type: 1 /* MultiChoice */
        };
        this.extraConfigurations.push(featureConfiguration);
    }
    setExtraConfiguration(extraConfigKeys) {
        this.userExtraConfiguration = extraConfigKeys;
    }
    generateFiles(projectPath, name, ...options) {
        const features = this.gridHelper.generateFeatures(this.userExtraConfiguration["features"], 3);
        const config = { "$(treeGridFeatures)": features };
        return super.generateFiles(projectPath, name, { extraConfig: config });
    }
    getExtraConfiguration() {
        return this.extraConfigurations;
    }
}
module.exports = new TreeGridExportTemplate();
