// igniteui-cli/migrations/update-3_2/index.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
// tslint:disable:no-implicit-dependencies
const schematics_1 = require("@angular-devkit/schematics");
// tslint:disable-next-line:no-submodule-imports
const testing_1 = require("@angular-devkit/schematics/testing");
describe("Update 3.2.0", () => {
    let appTree;
    const schematicRunner = new testing_1.SchematicTestRunner("ig-migrate", path.join(__dirname, "../migration-collection.json"));
    const configJson = {
        defaultProject: "testProj",
        projects: {
            testProj: {
                sourceRoot: "/src"
            }
        }
    };
    beforeEach(() => {
        appTree = new testing_1.UnitTestTree(new schematics_1.EmptyTree());
        appTree.create("/angular.json", JSON.stringify(configJson));
    });
    it("should update CustomDateSummary summaryResult assignment", done => {
        const summaryFile = "/src/app/grid-summaries/grid-summaries.component.ts";
        appTree.create(summaryFile, `class CustomDateSummary extends IgxDateSummaryOperand {
   constructor() {
     super();
   }
   public operate(data?: any[]): IgxSummaryResult[] {
     const result = [];
    result.push({
      key: 'earliest',
      label: 'Earliest Date',
      summaryResult: (IgxDateSummaryOperand.earliest(data)).toLocaleDateString()
    });
    result.push({
      key: 'latest',
      label: 'Latest Date',
      summaryResult: (IgxDateSummaryOperand.latest(data)).toLocaleDateString()
    });
     return result;
   }
 }`);
        schematicRunner.runSchematic("migration-03", {}, appTree);
        expect(appTree.readContent(summaryFile))
            .toEqual(`class CustomDateSummary extends IgxDateSummaryOperand {
   constructor() {
     super();
   }
   public operate(data?: any[]): IgxSummaryResult[] {
     const result = [];
    result.push({
      key: 'earliest',
      label: 'Earliest Date',
      summaryResult: data.length ? (IgxDateSummaryOperand.earliest(data)).toLocaleDateString() : null
    });
    result.push({
      key: 'latest',
      label: 'Latest Date',
      summaryResult: data.length ? (IgxDateSummaryOperand.latest(data)).toLocaleDateString() : null
    });
     return result;
   }
 }`);
        done();
    });
});
