// igniteui-cli/templates/angular/ig-ts/tree-grid/tree-grid-editing/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const GridHelper_1 = require("../../../../../lib/project-utility/GridHelper");
const AngularTemplate_1 = require("../../../../../lib/templates/AngularTemplate");
class TreeGridEditingTemplate extends AngularTemplate_1.AngularTemplate {
    constructor() {
        super(__dirname);
        this.userExtraConfiguration = {};
        this.id = "tree-grid-editing";
        this.name = "TreeGrid Editing";
        this.widget = "igTreeGrid";
        this.controlGroup = "Data Grids";
        this.description = "igTreeGrid editing template for Angular";
        this.dependencies = ["igTreeGrid"];
        this.projectType = "ig-ts";
        this.listInComponentTemplates = true;
        this.hasExtraConfiguration = true;
        this.extraConfigurations = [];
        this.gridHelper = new GridHelper_1.GridHelper();
        this.gridHelper.tree = true;
        const featureConfiguration = {
            choices: ["Sorting", "Filtering"],
            default: "",
            key: "features",
            message: "Select optional features for the editing template",
            type: 1 /* MultiChoice */
        };
        this.extraConfigurations.push(featureConfiguration);
    }
    setExtraConfiguration(extraConfigKeys) {
        this.userExtraConfiguration = extraConfigKeys;
    }
    generateFiles(projectPath, name, ...options) {
        this.gridHelper.addFeature("Updating", { columnSettings: [
                {
                    columnKey: "progress",
                    editorOptions: {
                        buttonType: "spin"
                    },
                    editorType: "currency"
                },
                {
                    columnKey: "start",
                    editorType: "datepicker"
                },
                {
                    columnKey: "finish",
                    editorType: "datepicker"
                }
            ] });
        const features = this.gridHelper.generateFeatures(this.userExtraConfiguration["features"], 3);
        const config = { "$(treeGridFeatures)": features };
        return super.generateFiles(projectPath, name, { extraConfig: config });
    }
    getExtraConfiguration() {
        return this.extraConfigurations;
    }
}
module.exports = new TreeGridEditingTemplate();
