// igniteui-cli/templates/angular/igx-ts/custom-templates/awesome-grid/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const IgniteUIForAngularTemplate_1 = require("../../../../../lib/templates/IgniteUIForAngularTemplate");
class IgxGridAwesomeTemplate extends IgniteUIForAngularTemplate_1.IgniteUIForAngularTemplate {
    constructor() {
        super(__dirname);
        this.id = "awesome-grid";
        this.projectType = "igx-ts";
        this.listInComponentTemplates = false;
        this.listInCustomTemplates = true;
        this.name = "Awesome Grid";
        this.description = "Awesome IgxGrid";
        this.dependencies = [
            { import: "IgxGridModule", from: "igniteui-angular" },
            {
                import: [
                    "IgxProgressBarModule",
                    "IgxIconModule",
                    "IgxAvatarModule",
                    "IgxBadgeModule",
                    "IgxSwitchModule",
                    "IgxInputGroupModule",
                    "IgxButtonModule"
                ],
                from: "igniteui-angular"
            },
            { import: "FormsModule", from: "@angular/forms" }
        ];
    }
}
module.exports = new IgxGridAwesomeTemplate();
