// igniteui-angular/esm5/lib/directives/drag-drop/drag-drop.strategy.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function IDropStrategy() { }
if (false) {
    /** @type {?} */
    IDropStrategy.prototype.dropAction;
}
// @dynamic
var 
// @dynamic
IgxDefaultDropStrategy = /** @class */ (function () {
    function IgxDefaultDropStrategy() {
    }
    /**
     * @param {?} drag
     * @param {?} drop
     * @param {?} atIndex
     * @return {?}
     */
    IgxDefaultDropStrategy.prototype.dropAction = /**
     * @param {?} drag
     * @param {?} drop
     * @param {?} atIndex
     * @return {?}
     */
    function (drag, drop, atIndex) { };
    return IgxDefaultDropStrategy;
}());
// @dynamic
export { IgxDefaultDropStrategy };
// @dynamic
var 
// @dynamic
IgxAppendDropStrategy = /** @class */ (function () {
    function IgxAppendDropStrategy(_renderer) {
        this._renderer = _renderer;
    }
    /**
     * @param {?} drag
     * @param {?} drop
     * @param {?} atIndex
     * @return {?}
     */
    IgxAppendDropStrategy.prototype.dropAction = /**
     * @param {?} drag
     * @param {?} drop
     * @param {?} atIndex
     * @return {?}
     */
    function (drag, drop, atIndex) {
        /** @type {?} */
        var dragElement = drag.element.nativeElement;
        /** @type {?} */
        var dropAreaElement = drop.element.nativeElement;
        this._renderer.removeChild(dragElement.parentNode, dragElement);
        this._renderer.appendChild(dropAreaElement, dragElement);
    };
    return IgxAppendDropStrategy;
}());
// @dynamic
export { IgxAppendDropStrategy };
if (false) {
    /**
     * @type {?}
     * @private
     */
    IgxAppendDropStrategy.prototype._renderer;
}
// @dynamic
var 
// @dynamic
IgxPrependDropStrategy = /** @class */ (function () {
    function IgxPrependDropStrategy(_renderer) {
        this._renderer = _renderer;
    }
    /**
     * @param {?} drag
     * @param {?} drop
     * @param {?} atIndex
     * @return {?}
     */
    IgxPrependDropStrategy.prototype.dropAction = /**
     * @param {?} drag
     * @param {?} drop
     * @param {?} atIndex
     * @return {?}
     */
    function (drag, drop, atIndex) {
        /** @type {?} */
        var dragElement = drag.element.nativeElement;
        /** @type {?} */
        var dropAreaElement = drop.element.nativeElement;
        this._renderer.removeChild(dragElement.parentNode, dragElement);
        if (dropAreaElement.children.length) {
            this._renderer.insertBefore(dropAreaElement, dragElement, dropAreaElement.children[0]);
        }
        else {
            this._renderer.appendChild(dropAreaElement, dragElement);
        }
    };
    return IgxPrependDropStrategy;
}());
// @dynamic
export { IgxPrependDropStrategy };
if (false) {
    /**
     * @type {?}
     * @private
     */
    IgxPrependDropStrategy.prototype._renderer;
}
// @dynamic
var 
// @dynamic
IgxInsertDropStrategy = /** @class */ (function () {
    function IgxInsertDropStrategy(_renderer) {
        this._renderer = _renderer;
    }
    /**
     * @param {?} drag
     * @param {?} drop
     * @param {?} atIndex
     * @return {?}
     */
    IgxInsertDropStrategy.prototype.dropAction = /**
     * @param {?} drag
     * @param {?} drop
     * @param {?} atIndex
     * @return {?}
     */
    function (drag, drop, atIndex) {
        if (drag.element.nativeElement.parentElement === drop.element.nativeElement && atIndex === -1) {
            return;
        }
        /** @type {?} */
        var dragElement = drag.element.nativeElement;
        /** @type {?} */
        var dropAreaElement = drop.element.nativeElement;
        this._renderer.removeChild(dragElement.parentNode, dragElement);
        if (atIndex !== -1 && dropAreaElement.children.length > atIndex) {
            this._renderer.insertBefore(dropAreaElement, dragElement, dropAreaElement.children[atIndex]);
        }
        else {
            this._renderer.appendChild(dropAreaElement, dragElement);
        }
    };
    return IgxInsertDropStrategy;
}());
// @dynamic
export { IgxInsertDropStrategy };
if (false) {
    /**
     * @type {?}
     * @private
     */
    IgxInsertDropStrategy.prototype._renderer;
}
