// igniteui-angular/esm5/lib/tabs/tabs.common.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @hidden
 * @abstract
 */
var /**
 * @hidden
 * @abstract
 */
IgxTabsBase = /** @class */ (function () {
    function IgxTabsBase() {
    }
    /**
     * @param {?} element
     * @param {?} scrollRight
     * @return {?}
     */
    IgxTabsBase.prototype.scrollElement = /**
     * @param {?} element
     * @param {?} scrollRight
     * @return {?}
     */
    function (element, scrollRight) { };
    /**
     * @param {?} newTab
     * @return {?}
     */
    IgxTabsBase.prototype.performSelectionChange = /**
     * @param {?} newTab
     * @return {?}
     */
    function (newTab) { };
    /**
     * @param {?} tab
     * @param {?} duration
     * @return {?}
     */
    IgxTabsBase.prototype.transformContentAnimation = /**
     * @param {?} tab
     * @param {?} duration
     * @return {?}
     */
    function (tab, duration) { };
    /**
     * @param {?} element
     * @param {?=} duration
     * @return {?}
     */
    IgxTabsBase.prototype.transformIndicatorAnimation = /**
     * @param {?} element
     * @param {?=} duration
     * @return {?}
     */
    function (element, duration) { };
    return IgxTabsBase;
}());
/**
 * @hidden
 * @abstract
 */
export { IgxTabsBase };
if (false) {
    /** @type {?} */
    IgxTabsBase.prototype.selectedIndicator;
    /** @type {?} */
    IgxTabsBase.prototype.tabs;
    /** @type {?} */
    IgxTabsBase.prototype.groups;
    /** @type {?} */
    IgxTabsBase.prototype.offset;
    /** @type {?} */
    IgxTabsBase.prototype.selectedIndex;
    /** @type {?} */
    IgxTabsBase.prototype.viewPort;
    /** @type {?} */
    IgxTabsBase.prototype.contentsContainer;
    /** @type {?} */
    IgxTabsBase.prototype.tabsContainer;
    /** @type {?} */
    IgxTabsBase.prototype.itemsContainer;
    /** @type {?} */
    IgxTabsBase.prototype.headerContainer;
    /** @type {?} */
    IgxTabsBase.prototype.onTabItemSelected;
    /** @type {?} */
    IgxTabsBase.prototype.hasContentTabs;
}
/**
 * @hidden
 * @abstract
 */
var /**
 * @hidden
 * @abstract
 */
IgxTabItemBase = /** @class */ (function () {
    function IgxTabItemBase() {
    }
    Object.defineProperty(IgxTabItemBase.prototype, "index", {
        get: /**
         * @return {?}
         */
        function () { return 0; },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    IgxTabItemBase.prototype.select = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} newValue
     * @return {?}
     */
    IgxTabItemBase.prototype.setSelectedInternal = /**
     * @param {?} newValue
     * @return {?}
     */
    function (newValue) { };
    return IgxTabItemBase;
}());
/**
 * @hidden
 * @abstract
 */
export { IgxTabItemBase };
if (false) {
    /** @type {?} */
    IgxTabItemBase.prototype.nativeTabItem;
}
/**
 * @hidden
 * @abstract
 */
var /**
 * @hidden
 * @abstract
 */
IgxTabsGroupBase = /** @class */ (function () {
    function IgxTabsGroupBase() {
    }
    /**
     * @return {?}
     */
    IgxTabsGroupBase.prototype.select = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} newValue
     * @return {?}
     */
    IgxTabsGroupBase.prototype.setSelectedInternal = /**
     * @param {?} newValue
     * @return {?}
     */
    function (newValue) { };
    return IgxTabsGroupBase;
}());
/**
 * @hidden
 * @abstract
 */
export { IgxTabsGroupBase };
