// igniteui-angular/esm2015/lib/directives/drag-drop/drag-drop.strategy.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function IDropStrategy() { }
if (false) {
    /** @type {?} */
    IDropStrategy.prototype.dropAction;
}
// @dynamic
export class IgxDefaultDropStrategy {
    /**
     * @param {?} drag
     * @param {?} drop
     * @param {?} atIndex
     * @return {?}
     */
    dropAction(drag, drop, atIndex) { }
}
// @dynamic
export class IgxAppendDropStrategy {
    /**
     * @param {?} _renderer
     */
    constructor(_renderer) {
        this._renderer = _renderer;
    }
    /**
     * @param {?} drag
     * @param {?} drop
     * @param {?} atIndex
     * @return {?}
     */
    dropAction(drag, drop, atIndex) {
        /** @type {?} */
        const dragElement = drag.element.nativeElement;
        /** @type {?} */
        const dropAreaElement = drop.element.nativeElement;
        this._renderer.removeChild(dragElement.parentNode, dragElement);
        this._renderer.appendChild(dropAreaElement, dragElement);
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    IgxAppendDropStrategy.prototype._renderer;
}
// @dynamic
export class IgxPrependDropStrategy {
    /**
     * @param {?} _renderer
     */
    constructor(_renderer) {
        this._renderer = _renderer;
    }
    /**
     * @param {?} drag
     * @param {?} drop
     * @param {?} atIndex
     * @return {?}
     */
    dropAction(drag, drop, atIndex) {
        /** @type {?} */
        const dragElement = drag.element.nativeElement;
        /** @type {?} */
        const dropAreaElement = drop.element.nativeElement;
        this._renderer.removeChild(dragElement.parentNode, dragElement);
        if (dropAreaElement.children.length) {
            this._renderer.insertBefore(dropAreaElement, dragElement, dropAreaElement.children[0]);
        }
        else {
            this._renderer.appendChild(dropAreaElement, dragElement);
        }
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    IgxPrependDropStrategy.prototype._renderer;
}
// @dynamic
export class IgxInsertDropStrategy {
    /**
     * @param {?} _renderer
     */
    constructor(_renderer) {
        this._renderer = _renderer;
    }
    /**
     * @param {?} drag
     * @param {?} drop
     * @param {?} atIndex
     * @return {?}
     */
    dropAction(drag, drop, atIndex) {
        if (drag.element.nativeElement.parentElement === drop.element.nativeElement && atIndex === -1) {
            return;
        }
        /** @type {?} */
        const dragElement = drag.element.nativeElement;
        /** @type {?} */
        const dropAreaElement = drop.element.nativeElement;
        this._renderer.removeChild(dragElement.parentNode, dragElement);
        if (atIndex !== -1 && dropAreaElement.children.length > atIndex) {
            this._renderer.insertBefore(dropAreaElement, dragElement, dropAreaElement.children[atIndex]);
        }
        else {
            this._renderer.appendChild(dropAreaElement, dragElement);
        }
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    IgxInsertDropStrategy.prototype._renderer;
}
