// igniteui-angular/esm5/lib/date-picker/date-picker.utils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { isIE } from '../core/utils';
/** @enum {string} */
var DateState = {
    Valid: 'valid',
    Invalid: 'invalid',
};
export { DateState };
/** @enum {string} */
var FormatDesc = {
    Numeric: 'numeric',
    TwoDigits: '2-digit',
};
/** @enum {string} */
var DateChars = {
    YearChar: 'y',
    MonthChar: 'M',
    DayChar: 'd',
};
/** @enum {string} */
var DateParts = {
    Day: 'day',
    Month: 'month',
    Year: 'year',
};
/**
 * @hidden
 * @abstract
 */
var DatePickerUtil = /** @class */ (function () {
    function DatePickerUtil() {
    }
    /**
     * This method generates date parts structure based on editor mask and locale.
     * @param maskValue: string
     * @param locale: string
     * @returns array containing information about date parts - type, position, format
     */
    /**
     * This method generates date parts structure based on editor mask and locale.
     * @param {?} maskValue
     * @param {?=} locale
     * @return {?} array containing information about date parts - type, position, format
     */
    DatePickerUtil.parseDateFormat = /**
     * This method generates date parts structure based on editor mask and locale.
     * @param {?} maskValue
     * @param {?=} locale
     * @return {?} array containing information about date parts - type, position, format
     */
    function (maskValue, locale) {
        if (locale === void 0) { locale = DatePickerUtil.DEFAULT_LOCALE; }
        /** @type {?} */
        var dateStruct = [];
        if (maskValue === undefined && !isIE()) {
            dateStruct = DatePickerUtil.getDefaultLocaleMask(locale);
        }
        else {
            /** @type {?} */
            var mask = (maskValue) ? maskValue : DatePickerUtil.SHORT_DATE_MASK;
            /** @type {?} */
            var maskArray = Array.from(mask);
            /** @type {?} */
            var monthInitPosition = mask.indexOf("M" /* MonthChar */);
            /** @type {?} */
            var dayInitPosition = mask.indexOf("d" /* DayChar */);
            /** @type {?} */
            var yearInitPosition = mask.indexOf("y" /* YearChar */);
            if (yearInitPosition !== -1) {
                dateStruct.push({
                    type: "year" /* Year */,
                    initialPosition: yearInitPosition,
                    formatType: DatePickerUtil.getYearFormatType(mask)
                });
            }
            if (monthInitPosition !== -1) {
                dateStruct.push({
                    type: "month" /* Month */,
                    initialPosition: monthInitPosition,
                    formatType: DatePickerUtil.getMonthFormatType(mask)
                });
            }
            if (dayInitPosition !== -1) {
                dateStruct.push({
                    type: "day" /* Day */,
                    initialPosition: dayInitPosition,
                    formatType: DatePickerUtil.getDayFormatType(mask)
                });
            }
            for (var i = 0; i < maskArray.length; i++) {
                if (!DatePickerUtil.isDateChar(maskArray[i])) {
                    dateStruct.push({
                        type: DatePickerUtil.SEPARATOR,
                        initialPosition: i,
                        value: maskArray[i]
                    });
                }
            }
            dateStruct.sort(function (a, b) { return a.initialPosition - b.initialPosition; });
            DatePickerUtil.fillDatePartsPositions(dateStruct);
        }
        return dateStruct;
    };
    /**
     * This method generates input mask based on date parts.
     * @param dateStruct array
     * @returns input mask
     */
    /**
     * This method generates input mask based on date parts.
     * @param {?} dateStruct array
     * @return {?} input mask
     */
    DatePickerUtil.getInputMask = /**
     * This method generates input mask based on date parts.
     * @param {?} dateStruct array
     * @return {?} input mask
     */
    function (dateStruct) {
        /** @type {?} */
        var inputMask = [];
        for (var i = 0; i < dateStruct.length; i++) {
            if (dateStruct[i].type === DatePickerUtil.SEPARATOR) {
                inputMask.push(dateStruct[i].value);
            }
            else if (dateStruct[i].type === "day" /* Day */ || dateStruct[i].type === "month" /* Month */) {
                inputMask.push('00');
            }
            else if (dateStruct[i].type === "year" /* Year */) {
                switch (dateStruct[i].formatType) {
                    case "numeric" /* Numeric */: {
                        inputMask.push('0000');
                        break;
                    }
                    case "2-digit" /* TwoDigits */: {
                        inputMask.push('00');
                        break;
                    }
                }
            }
        }
        return inputMask.join('');
    };
    /**
     * This method generates editor mask.
     * @param dateStruct
     * @returns editor mask
     */
    /**
     * This method generates editor mask.
     * @param {?} dateStruct
     * @return {?} editor mask
     */
    DatePickerUtil.getMask = /**
     * This method generates editor mask.
     * @param {?} dateStruct
     * @return {?} editor mask
     */
    function (dateStruct) {
        /** @type {?} */
        var mask = [];
        for (var i = 0; i < dateStruct.length; i++) {
            switch (dateStruct[i].formatType) {
                case "numeric" /* Numeric */: {
                    if (dateStruct[i].type === "day" /* Day */) {
                        mask.push('d');
                    }
                    else if (dateStruct[i].type === "month" /* Month */) {
                        mask.push('M');
                    }
                    else {
                        mask.push('yyyy');
                    }
                    break;
                }
                case "2-digit" /* TwoDigits */: {
                    if (dateStruct[i].type === "day" /* Day */) {
                        mask.push('dd');
                    }
                    else if (dateStruct[i].type === "month" /* Month */) {
                        mask.push('MM');
                    }
                    else {
                        mask.push('yy');
                    }
                }
            }
            if (dateStruct[i].type === DatePickerUtil.SEPARATOR) {
                mask.push(dateStruct[i].value);
            }
        }
        return mask.join('');
    };
    /**
     * This method parses an input string base on date parts and returns a date and its validation state.
     * @param dateFormatParts
     * @param prevDateValue
     * @param inputValue
     * @returns object containing a date and its validation state
     */
    /**
     * This method parses an input string base on date parts and returns a date and its validation state.
     * @param {?} dateFormatParts
     * @param {?} prevDateValue
     * @param {?} inputValue
     * @return {?} object containing a date and its validation state
     */
    DatePickerUtil.parseDateArray = /**
     * This method parses an input string base on date parts and returns a date and its validation state.
     * @param {?} dateFormatParts
     * @param {?} prevDateValue
     * @param {?} inputValue
     * @return {?} object containing a date and its validation state
     */
    function (dateFormatParts, prevDateValue, inputValue) {
        /** @type {?} */
        var dayStr = DatePickerUtil.getDayValueFromInput(dateFormatParts, inputValue);
        /** @type {?} */
        var monthStr = DatePickerUtil.getMonthValueFromInput(dateFormatParts, inputValue);
        /** @type {?} */
        var yearStr = DatePickerUtil.getYearValueFromInput(dateFormatParts, inputValue);
        /** @type {?} */
        var yearFormat = DatePickerUtil.getDateFormatPart(dateFormatParts, "year" /* Year */).formatType;
        /** @type {?} */
        var day = (dayStr !== '') ? parseInt(dayStr, 10) : 1;
        /** @type {?} */
        var month = (monthStr !== '') ? parseInt(monthStr, 10) - 1 : 0;
        /** @type {?} */
        var year;
        if (yearStr === '') {
            year = (yearFormat === "2-digit" /* TwoDigits */) ? '00' : '2000';
        }
        else {
            year = yearStr;
        }
        /** @type {?} */
        var yearPrefix;
        if (prevDateValue) {
            /** @type {?} */
            var originalYear = prevDateValue.getFullYear().toString();
            if (originalYear.length === 4) {
                yearPrefix = originalYear.substring(0, 2);
            }
        }
        else {
            yearPrefix = '20';
        }
        /** @type {?} */
        var fullYear = (yearFormat === "2-digit" /* TwoDigits */) ? yearPrefix.concat(year) : year;
        if ((month < 0) || (month > 11) || (month === NaN)) {
            return { state: "invalid" /* Invalid */, value: inputValue };
        }
        if ((day < 1) || (day > DatePickerUtil.daysInMonth(fullYear, month + 1)) || (day === NaN)) {
            return { state: "invalid" /* Invalid */, value: inputValue };
        }
        return { state: "valid" /* Valid */, date: new Date(fullYear, month, day) };
    };
    /**
     * @param {?} mask
     * @return {?}
     */
    DatePickerUtil.maskToPromptChars = /**
     * @param {?} mask
     * @return {?}
     */
    function (mask) {
        /** @type {?} */
        var result = mask.replace(/0|L/g, DatePickerUtil.PROMPT_CHAR);
        return result;
    };
    /**
     * This method replaces prompt chars with empty string.
     * @param value
     */
    /**
     * This method replaces prompt chars with empty string.
     * @param {?} value
     * @return {?}
     */
    DatePickerUtil.trimUnderlines = /**
     * This method replaces prompt chars with empty string.
     * @param {?} value
     * @return {?}
     */
    function (value) {
        /** @type {?} */
        var result = value.replace(/_/g, '');
        return result;
    };
    /**
     * This method is used for spinning date parts.
     * @param dateFormatParts
     * @param inputValue
     * @param position
     * @param delta
     * @param isSpinLoop
     * @return modified text input
     */
    /**
     * This method is used for spinning date parts.
     * @param {?} dateFormatParts
     * @param {?} inputValue
     * @param {?} position
     * @param {?} delta
     * @param {?} isSpinLoop
     * @return {?} modified text input
     */
    DatePickerUtil.getModifiedDateInput = /**
     * This method is used for spinning date parts.
     * @param {?} dateFormatParts
     * @param {?} inputValue
     * @param {?} position
     * @param {?} delta
     * @param {?} isSpinLoop
     * @return {?} modified text input
     */
    function (dateFormatParts, inputValue, position, delta, isSpinLoop) {
        /** @type {?} */
        var datePart = DatePickerUtil.getDatePartOnPosition(dateFormatParts, position);
        /** @type {?} */
        var datePartType = datePart.type;
        /** @type {?} */
        var datePartFormatType = datePart.formatType;
        /** @type {?} */
        var newValue;
        /** @type {?} */
        var datePartValue = DatePickerUtil.getDateValueFromInput(dateFormatParts, datePartType, inputValue);
        newValue = parseInt(datePartValue, 10);
        /** @type {?} */
        var maxValue;
        /** @type {?} */
        var minValue;
        /** @type {?} */
        var minMax = DatePickerUtil.getMinMaxValue(dateFormatParts, datePart, inputValue);
        minValue = minMax.min;
        maxValue = minMax.max;
        if (isNaN(newValue)) {
            if (minValue === 'infinite') {
                newValue = 2000;
            }
            else {
                newValue = minValue;
            }
        }
        /** @type {?} */
        var tempValue = newValue;
        tempValue += delta;
        // Infinite loop for full years
        if (maxValue === 'infinite' && minValue === 'infinite') {
            newValue = tempValue;
        }
        if (isSpinLoop) {
            if (tempValue > maxValue) {
                tempValue = minValue;
            }
            if (tempValue < minValue) {
                tempValue = maxValue;
            }
            newValue = tempValue;
        }
        else {
            if (tempValue <= maxValue && tempValue >= minValue) {
                newValue = tempValue;
            }
        }
        /** @type {?} */
        var startIdx = datePart.position[0];
        /** @type {?} */
        var endIdx = datePart.position[1];
        /** @type {?} */
        var start = inputValue.slice(0, startIdx);
        /** @type {?} */
        var end = inputValue.slice(endIdx, inputValue.length);
        /** @type {?} */
        var changedPart;
        /** @type {?} */
        var prefix = DatePickerUtil.getNumericFormatPrefix(datePartFormatType);
        changedPart = (newValue < 10) ? "" + prefix + newValue : "" + newValue;
        return "" + start + changedPart + end;
    };
    /**
     * This method returns date input with prompt chars.
     * @param dateFormatParts
     * @param date
     * @param inputValue
     * @returns date input including prompt chars
     */
    /**
     * This method returns date input with prompt chars.
     * @param {?} dateFormatParts
     * @param {?} date
     * @param {?} inputValue
     * @return {?} date input including prompt chars
     */
    DatePickerUtil.addPromptCharsEditMode = /**
     * This method returns date input with prompt chars.
     * @param {?} dateFormatParts
     * @param {?} date
     * @param {?} inputValue
     * @return {?} date input including prompt chars
     */
    function (dateFormatParts, date, inputValue) {
        /** @type {?} */
        var dateArray = Array.from(inputValue);
        for (var i = 0; i < dateFormatParts.length; i++) {
            if (dateFormatParts[i].formatType === "numeric" /* Numeric */) {
                if ((dateFormatParts[i].type === "day" /* Day */ && date.getDate() < 10)
                    || (dateFormatParts[i].type === "month" /* Month */ && date.getMonth() + 1 < 10)) {
                    dateArray.splice(dateFormatParts[i].position[0], 0, DatePickerUtil.PROMPT_CHAR);
                    dateArray.join('');
                }
            }
        }
        return dateArray.join('');
    };
    /**
     * This method checks if date input is done.
     * @param dateFormatParts
     * @param input
     * @returns input completeness
     */
    /**
     * This method checks if date input is done.
     * @param {?} dateFormatParts
     * @param {?} input
     * @return {?} input completeness
     */
    DatePickerUtil.checkForCompleteDateInput = /**
     * This method checks if date input is done.
     * @param {?} dateFormatParts
     * @param {?} input
     * @return {?} input completeness
     */
    function (dateFormatParts, input) {
        /** @type {?} */
        var dayValue = DatePickerUtil.getDayValueFromInput(dateFormatParts, input);
        /** @type {?} */
        var monthValue = DatePickerUtil.getMonthValueFromInput(dateFormatParts, input);
        /** @type {?} */
        var yearValue = DatePickerUtil.getYearValueFromInput(dateFormatParts, input);
        /** @type {?} */
        var dayStr = DatePickerUtil.getDayValueFromInput(dateFormatParts, input, false);
        /** @type {?} */
        var monthStr = DatePickerUtil.getMonthValueFromInput(dateFormatParts, input, false);
        if (DatePickerUtil.isFullInput(dayValue, dayStr)
            && DatePickerUtil.isFullInput(monthValue, monthStr)
            && DatePickerUtil.isFullYearInput(dateFormatParts, yearValue)) {
            return 'complete';
        }
        else if (dayValue === '' && monthValue === '' && yearValue === '') {
            return 'empty';
        }
        else if (dayValue === '' || monthValue === '' || yearValue === '') {
            return 'partial';
        }
        return '';
    };
    /**
     * @private
     * @param {?} format
     * @return {?}
     */
    DatePickerUtil.getYearFormatType = /**
     * @private
     * @param {?} format
     * @return {?}
     */
    function (format) {
        switch (format.match(new RegExp("y" /* YearChar */, 'g')).length) {
            case 1: {
                // y (2020)
                return "numeric" /* Numeric */;
            }
            case 4: {
                // yyyy (2020)
                return "numeric" /* Numeric */;
            }
            case 2: {
                // yy (20)
                return "2-digit" /* TwoDigits */;
            }
        }
    };
    /**
     * @private
     * @param {?} format
     * @return {?}
     */
    DatePickerUtil.getMonthFormatType = /**
     * @private
     * @param {?} format
     * @return {?}
     */
    function (format) {
        switch (format.match(new RegExp("M" /* MonthChar */, 'g')).length) {
            case 1: {
                // M (8)
                return "numeric" /* Numeric */;
            }
            case 2: {
                // MM (08)
                return "2-digit" /* TwoDigits */;
            }
        }
    };
    /**
     * @private
     * @param {?} format
     * @return {?}
     */
    DatePickerUtil.getDayFormatType = /**
     * @private
     * @param {?} format
     * @return {?}
     */
    function (format) {
        switch (format.match(new RegExp("d" /* DayChar */, 'g')).length) {
            case 1: {
                // d (6)
                return "numeric" /* Numeric */;
            }
            case 2: {
                // dd (06)
                return "2-digit" /* TwoDigits */;
            }
        }
    };
    /**
     * @private
     * @param {?} locale
     * @return {?}
     */
    DatePickerUtil.getDefaultLocaleMask = /**
     * @private
     * @param {?} locale
     * @return {?}
     */
    function (locale) {
        /** @type {?} */
        var dateStruct = [];
        /** @type {?} */
        var formatter = new Intl.DateTimeFormat(locale);
        /** @type {?} */
        var formatToParts = formatter.formatToParts(new Date());
        for (var i = 0; i < formatToParts.length; i++) {
            if (formatToParts[i].type === DatePickerUtil.SEPARATOR) {
                dateStruct.push({
                    type: DatePickerUtil.SEPARATOR,
                    value: formatToParts[i].value
                });
            }
            else {
                dateStruct.push({
                    type: formatToParts[i].type,
                });
            }
        }
        /** @type {?} */
        var formatterOptions = formatter.resolvedOptions();
        for (var i = 0; i < dateStruct.length; i++) {
            switch (dateStruct[i].type) {
                case "day" /* Day */: {
                    dateStruct[i].formatType = formatterOptions.day;
                    break;
                }
                case "month" /* Month */: {
                    dateStruct[i].formatType = formatterOptions.month;
                    break;
                }
                case "year" /* Year */: {
                    dateStruct[i].formatType = formatterOptions.month;
                    break;
                }
            }
        }
        DatePickerUtil.fillDatePartsPositions(dateStruct);
        return dateStruct;
    };
    /**
     * @private
     * @param {?} char
     * @return {?}
     */
    DatePickerUtil.isDateChar = /**
     * @private
     * @param {?} char
     * @return {?}
     */
    function (char) {
        return (char === "y" /* YearChar */ || char === "M" /* MonthChar */ || char === "d" /* DayChar */);
    };
    /**
     * @private
     * @param {?} formatType
     * @return {?}
     */
    DatePickerUtil.getNumericFormatPrefix = /**
     * @private
     * @param {?} formatType
     * @return {?}
     */
    function (formatType) {
        switch (formatType) {
            case "2-digit" /* TwoDigits */: {
                return '0';
            }
            case "numeric" /* Numeric */: {
                return DatePickerUtil.PROMPT_CHAR;
            }
        }
    };
    /**
     * @private
     * @param {?} dateFormatParts
     * @param {?} datePart
     * @param {?} inputValue
     * @return {?}
     */
    DatePickerUtil.getMinMaxValue = /**
     * @private
     * @param {?} dateFormatParts
     * @param {?} datePart
     * @param {?} inputValue
     * @return {?}
     */
    function (dateFormatParts, datePart, inputValue) {
        /** @type {?} */
        var maxValue;
        /** @type {?} */
        var minValue;
        switch (datePart.type) {
            case "month" /* Month */: {
                minValue = 1;
                maxValue = DatePickerUtil.NUMBER_OF_MONTHS;
                break;
            }
            case "day" /* Day */: {
                minValue = 1;
                maxValue = DatePickerUtil.daysInMonth(DatePickerUtil.getFullYearFromString(DatePickerUtil.getDateFormatPart(dateFormatParts, "year" /* Year */), inputValue), parseInt(DatePickerUtil.getMonthValueFromInput(dateFormatParts, inputValue), 10));
                break;
            }
            case "year" /* Year */: {
                if (datePart.formatType === "2-digit" /* TwoDigits */) {
                    minValue = 0;
                    maxValue = 99;
                }
                else {
                    // Infinite loop
                    minValue = 'infinite';
                    maxValue = 'infinite';
                }
                break;
            }
        }
        return { min: minValue, max: maxValue };
    };
    /**
     * @private
     * @param {?} fullYear
     * @param {?} month
     * @return {?}
     */
    DatePickerUtil.daysInMonth = /**
     * @private
     * @param {?} fullYear
     * @param {?} month
     * @return {?}
     */
    function (fullYear, month) {
        return new Date(fullYear, month, 0).getDate();
    };
    /**
     * @private
     * @param {?} dateFormatParts
     * @param {?} type
     * @param {?} inputValue
     * @param {?=} trim
     * @return {?}
     */
    DatePickerUtil.getDateValueFromInput = /**
     * @private
     * @param {?} dateFormatParts
     * @param {?} type
     * @param {?} inputValue
     * @param {?=} trim
     * @return {?}
     */
    function (dateFormatParts, type, inputValue, trim) {
        if (trim === void 0) { trim = true; }
        /** @type {?} */
        var partPosition = DatePickerUtil.getDateFormatPart(dateFormatParts, type).position;
        /** @type {?} */
        var result = inputValue.substring(partPosition[0], partPosition[1]);
        return (trim) ? DatePickerUtil.trimUnderlines(result) : result;
    };
    /**
     * @private
     * @param {?} dateFormatParts
     * @param {?} inputValue
     * @param {?=} trim
     * @return {?}
     */
    DatePickerUtil.getDayValueFromInput = /**
     * @private
     * @param {?} dateFormatParts
     * @param {?} inputValue
     * @param {?=} trim
     * @return {?}
     */
    function (dateFormatParts, inputValue, trim) {
        if (trim === void 0) { trim = true; }
        return DatePickerUtil.getDateValueFromInput(dateFormatParts, "day" /* Day */, inputValue, trim);
    };
    /**
     * @private
     * @param {?} dateFormatParts
     * @param {?} inputValue
     * @param {?=} trim
     * @return {?}
     */
    DatePickerUtil.getMonthValueFromInput = /**
     * @private
     * @param {?} dateFormatParts
     * @param {?} inputValue
     * @param {?=} trim
     * @return {?}
     */
    function (dateFormatParts, inputValue, trim) {
        if (trim === void 0) { trim = true; }
        return DatePickerUtil.getDateValueFromInput(dateFormatParts, "month" /* Month */, inputValue, trim);
    };
    /**
     * @private
     * @param {?} dateFormatParts
     * @param {?} inputValue
     * @param {?=} trim
     * @return {?}
     */
    DatePickerUtil.getYearValueFromInput = /**
     * @private
     * @param {?} dateFormatParts
     * @param {?} inputValue
     * @param {?=} trim
     * @return {?}
     */
    function (dateFormatParts, inputValue, trim) {
        if (trim === void 0) { trim = true; }
        return DatePickerUtil.getDateValueFromInput(dateFormatParts, "year" /* Year */, inputValue, trim);
    };
    /**
     * @private
     * @param {?} dateFormatParts
     * @param {?} type
     * @return {?}
     */
    DatePickerUtil.getDateFormatPart = /**
     * @private
     * @param {?} dateFormatParts
     * @param {?} type
     * @return {?}
     */
    function (dateFormatParts, type) {
        /** @type {?} */
        var result = dateFormatParts.filter(function (datePart) { return (datePart.type === type); })[0];
        return result;
    };
    /**
     * @private
     * @param {?} value
     * @param {?} input
     * @return {?}
     */
    DatePickerUtil.isFullInput = /**
     * @private
     * @param {?} value
     * @param {?} input
     * @return {?}
     */
    function (value, input) {
        return (value !== '' && input.length === 2 && input.charAt(1) !== DatePickerUtil.PROMPT_CHAR);
    };
    /**
     * @private
     * @param {?} dateFormatParts
     * @param {?} value
     * @return {?}
     */
    DatePickerUtil.isFullYearInput = /**
     * @private
     * @param {?} dateFormatParts
     * @param {?} value
     * @return {?}
     */
    function (dateFormatParts, value) {
        switch (DatePickerUtil.getDateFormatPart(dateFormatParts, "year" /* Year */).formatType) {
            case "numeric" /* Numeric */: {
                return (value !== '' && value.length === 4);
            }
            case "2-digit" /* TwoDigits */: {
                return (value !== '' && value.length === 2);
            }
            default: {
                return false;
            }
        }
    };
    /**
     * @private
     * @param {?} dateFormatParts
     * @param {?} position
     * @return {?}
     */
    DatePickerUtil.getDatePartOnPosition = /**
     * @private
     * @param {?} dateFormatParts
     * @param {?} position
     * @return {?}
     */
    function (dateFormatParts, position) {
        /** @type {?} */
        var result = dateFormatParts.filter(function (element) {
            return element.position[0] <= position && position <= element.position[1] && element.type !== DatePickerUtil.SEPARATOR;
        })[0];
        return result;
    };
    /**
     * @private
     * @param {?} yearPart
     * @param {?} inputValue
     * @return {?}
     */
    DatePickerUtil.getFullYearFromString = /**
     * @private
     * @param {?} yearPart
     * @param {?} inputValue
     * @return {?}
     */
    function (yearPart, inputValue) {
        return parseInt(inputValue.substring(yearPart.position[0], yearPart.position[1]), 10);
    };
    /**
     * @private
     * @param {?} dateArray
     * @return {?}
     */
    DatePickerUtil.fillDatePartsPositions = /**
     * @private
     * @param {?} dateArray
     * @return {?}
     */
    function (dateArray) {
        /** @type {?} */
        var currentPos = 0;
        for (var i = 0; i < dateArray.length; i++) {
            // Day|Month part positions
            if (dateArray[i].type === "day" /* Day */ || dateArray[i].type === "month" /* Month */) {
                // Offset 2 positions for number
                dateArray[i].position = [currentPos, currentPos + 2];
                currentPos += 2;
            }
            else if (dateArray[i].type === "year" /* Year */) {
                // Year part positions
                switch (dateArray[i].formatType) {
                    case "numeric" /* Numeric */: {
                        // Offset 4 positions for full year
                        dateArray[i].position = [currentPos, currentPos + 4];
                        currentPos += 4;
                        break;
                    }
                    case "2-digit" /* TwoDigits */: {
                        // Offset 2 positions for short year
                        dateArray[i].position = [currentPos, currentPos + 2];
                        currentPos += 2;
                        break;
                    }
                }
            }
            else if (dateArray[i].type === DatePickerUtil.SEPARATOR) {
                // Separator positions
                dateArray[i].position = [currentPos, currentPos + 1];
                currentPos++;
            }
        }
    };
    DatePickerUtil.SHORT_DATE_MASK = 'MM/dd/yy';
    DatePickerUtil.SEPARATOR = 'literal';
    DatePickerUtil.NUMBER_OF_MONTHS = 12;
    DatePickerUtil.PROMPT_CHAR = '_';
    DatePickerUtil.DEFAULT_LOCALE = 'en';
    return DatePickerUtil;
}());
export { DatePickerUtil };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DatePickerUtil.SHORT_DATE_MASK;
    /**
     * @type {?}
     * @private
     */
    DatePickerUtil.SEPARATOR;
    /**
     * @type {?}
     * @private
     */
    DatePickerUtil.NUMBER_OF_MONTHS;
    /**
     * @type {?}
     * @private
     */
    DatePickerUtil.PROMPT_CHAR;
    /**
     * @type {?}
     * @private
     */
    DatePickerUtil.DEFAULT_LOCALE;
}
