// igniteui-angular/esm2015/lib/services/excel/excel-enums.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
const ExcelFolderTypes = {
    RootExcelFolder: 0,
    RootRelsExcelFolder: 1,
    DocPropsExcelFolder: 2,
    XLExcelFolder: 3,
    XLRelsExcelFolder: 4,
    ThemeExcelFolder: 5,
    WorksheetsExcelFolder: 6,
    WorksheetsRelsExcelFolder: 7,
    TablesExcelFolder: 8,
};
export { ExcelFolderTypes };
ExcelFolderTypes[ExcelFolderTypes.RootExcelFolder] = 'RootExcelFolder';
ExcelFolderTypes[ExcelFolderTypes.RootRelsExcelFolder] = 'RootRelsExcelFolder';
ExcelFolderTypes[ExcelFolderTypes.DocPropsExcelFolder] = 'DocPropsExcelFolder';
ExcelFolderTypes[ExcelFolderTypes.XLExcelFolder] = 'XLExcelFolder';
ExcelFolderTypes[ExcelFolderTypes.XLRelsExcelFolder] = 'XLRelsExcelFolder';
ExcelFolderTypes[ExcelFolderTypes.ThemeExcelFolder] = 'ThemeExcelFolder';
ExcelFolderTypes[ExcelFolderTypes.WorksheetsExcelFolder] = 'WorksheetsExcelFolder';
ExcelFolderTypes[ExcelFolderTypes.WorksheetsRelsExcelFolder] = 'WorksheetsRelsExcelFolder';
ExcelFolderTypes[ExcelFolderTypes.TablesExcelFolder] = 'TablesExcelFolder';
/** @enum {number} */
const ExcelFileTypes = {
    RootRelsFile: 0,
    AppFile: 1,
    CoreFile: 2,
    WorkbookRelsFile: 3,
    ThemeFile: 4,
    WorksheetFile: 5,
    StyleFile: 6,
    WorkbookFile: 7,
    ContentTypesFile: 8,
    SharedStringsFile: 9,
    WorksheetRelsFile: 10,
    TablesFile: 11,
};
export { ExcelFileTypes };
ExcelFileTypes[ExcelFileTypes.RootRelsFile] = 'RootRelsFile';
ExcelFileTypes[ExcelFileTypes.AppFile] = 'AppFile';
ExcelFileTypes[ExcelFileTypes.CoreFile] = 'CoreFile';
ExcelFileTypes[ExcelFileTypes.WorkbookRelsFile] = 'WorkbookRelsFile';
ExcelFileTypes[ExcelFileTypes.ThemeFile] = 'ThemeFile';
ExcelFileTypes[ExcelFileTypes.WorksheetFile] = 'WorksheetFile';
ExcelFileTypes[ExcelFileTypes.StyleFile] = 'StyleFile';
ExcelFileTypes[ExcelFileTypes.WorkbookFile] = 'WorkbookFile';
ExcelFileTypes[ExcelFileTypes.ContentTypesFile] = 'ContentTypesFile';
ExcelFileTypes[ExcelFileTypes.SharedStringsFile] = 'SharedStringsFile';
ExcelFileTypes[ExcelFileTypes.WorksheetRelsFile] = 'WorksheetRelsFile';
ExcelFileTypes[ExcelFileTypes.TablesFile] = 'TablesFile';
