// igniteui-cli/templates/react/es6/grid/grid-templating/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
//TODO:
const GridHelper_1 = require("../../../../../lib/project-utility/GridHelper");
const ReactTemplate_1 = require("../../../../../lib/templates/ReactTemplate");
class GridTemplatingTemplate extends ReactTemplate_1.ReactTemplate {
    /**
     *
     */
    constructor() {
        super(__dirname);
        this.extraConfigurations = [];
        this.userExtraConfiguration = {};
        this.id = "grid-templating";
        this.widget = "igGrid";
        this.name = "Grid Templating";
        this.description = "igGrid with templating for React";
        this.projectType = "es6";
        this.components = ["Grid"];
        this.controlGroup = "Data Grids";
        this.dependencies = ["igGrid"];
        this.gridHelper = new GridHelper_1.GridHelper();
        this.hasExtraConfiguration = true;
        this.extraConfigurations.push({
            choices: ["Sorting", "Paging", "Filtering"],
            default: "",
            key: "features",
            message: "Select features for the igGrid",
            type: 1 /* MultiChoice */
        });
    }
    generateFiles(projectPath, name, ...options) {
        const features = this.gridHelper.generateFeatures(this.userExtraConfiguration["features"], 5);
        const config = { "$(gridfeatures)": features };
        return super.generateFiles(projectPath, name, { extraConfig: config });
    }
    getExtraConfiguration() {
        return this.extraConfigurations;
    }
    setExtraConfiguration(extraConfigKeys) {
        this.userExtraConfiguration = extraConfigKeys;
    }
}
module.exports = new GridTemplatingTemplate();
