// igniteui-angular/esm5/lib/time-picker/time-picker.common.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @hidden
 * @type {?}
 */
export var IGX_TIME_PICKER_COMPONENT = 'IgxTimePickerComponentToken';
/**
 * @hidden
 * @record
 */
export function IgxTimePickerBase() { }
if (false) {
    /** @type {?} */
    IgxTimePickerBase.prototype.hourList;
    /** @type {?} */
    IgxTimePickerBase.prototype.minuteList;
    /** @type {?} */
    IgxTimePickerBase.prototype.ampmList;
    /** @type {?} */
    IgxTimePickerBase.prototype.selectedHour;
    /** @type {?} */
    IgxTimePickerBase.prototype.selectedMinute;
    /** @type {?} */
    IgxTimePickerBase.prototype.selectedAmPm;
    /** @type {?} */
    IgxTimePickerBase.prototype.format;
    /** @type {?} */
    IgxTimePickerBase.prototype.promptChar;
    /** @type {?} */
    IgxTimePickerBase.prototype.cleared;
    /** @type {?} */
    IgxTimePickerBase.prototype.mode;
    /** @type {?} */
    IgxTimePickerBase.prototype.showHoursList;
    /** @type {?} */
    IgxTimePickerBase.prototype.showMinutesList;
    /** @type {?} */
    IgxTimePickerBase.prototype.showAmPmList;
    /**
     * @return {?}
     */
    IgxTimePickerBase.prototype.nextHour = function () { };
    /**
     * @return {?}
     */
    IgxTimePickerBase.prototype.prevHour = function () { };
    /**
     * @return {?}
     */
    IgxTimePickerBase.prototype.nextMinute = function () { };
    /**
     * @return {?}
     */
    IgxTimePickerBase.prototype.prevMinute = function () { };
    /**
     * @return {?}
     */
    IgxTimePickerBase.prototype.nextAmPm = function () { };
    /**
     * @return {?}
     */
    IgxTimePickerBase.prototype.prevAmPm = function () { };
    /**
     * @return {?}
     */
    IgxTimePickerBase.prototype.okButtonClick = function () { };
    /**
     * @return {?}
     */
    IgxTimePickerBase.prototype.cancelButtonClick = function () { };
    /**
     * @param {?} item
     * @return {?}
     */
    IgxTimePickerBase.prototype.scrollHourIntoView = function (item) { };
    /**
     * @param {?} item
     * @return {?}
     */
    IgxTimePickerBase.prototype.scrollMinuteIntoView = function (item) { };
    /**
     * @param {?} item
     * @return {?}
     */
    IgxTimePickerBase.prototype.scrollAmPmIntoView = function (item) { };
    /**
     * @return {?}
     */
    IgxTimePickerBase.prototype.close = function () { };
    /**
     * @param {?=} preserveAmPm
     * @return {?}
     */
    IgxTimePickerBase.prototype.parseMask = function (preserveAmPm) { };
}
